<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Harmonise the alleles and effects between the exposure and outcome — harmonise_data • TwoSampleMR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Harmonise the alleles and effects between the exposure and outcome — harmonise_data"><meta name="description" content="In order to perform MR the effect of a SNP on an outcome and exposure must be harmonised to be
relative to the same allele."><meta property="og:description" content="In order to perform MR the effect of a SNP on an outcome and exposure must be harmonised to be
relative to the same allele."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide"><li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Harmonise the alleles and effects between the exposure and outcome</h1>
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/v0.6.12/R/harmonise.R" class="external-link"><code>R/harmonise.R</code></a></small>
      <div class="d-none name"><code>harmonise_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>In order to perform MR the effect of a SNP on an outcome and exposure must be harmonised to be
relative to the same allele.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">harmonise_data</span><span class="op">(</span><span class="va">exposure_dat</span>, <span class="va">outcome_dat</span>, action <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-exposure-dat">exposure_dat<a class="anchor" aria-label="anchor" href="#arg-exposure-dat"></a></dt>
<dd><p>Output from <code><a href="read_exposure_data.html">read_exposure_data()</a></code>.</p></dd>


<dt id="arg-outcome-dat">outcome_dat<a class="anchor" aria-label="anchor" href="#arg-outcome-dat"></a></dt>
<dd><p>Output from <code><a href="extract_outcome_data.html">extract_outcome_data()</a></code>.</p></dd>


<dt id="arg-action">action<a class="anchor" aria-label="anchor" href="#arg-action"></a></dt>
<dd><p>Level of strictness in dealing with SNPs.</p><ul><li><p><code>action = 1</code>: Assume all alleles are coded on the forward strand, i.e. do not attempt to flip alleles</p></li>
<li><p><code>action = 2</code>: Try to infer positive strand alleles, using allele frequencies for palindromes (default, conservative);</p></li>
<li><p><code>action = 3</code>: Correct strand for non-palindromic SNPs, and drop all palindromic SNPs from the analysis (more conservative).
If a single value is passed then this action is applied to all outcomes.
But multiple values can be supplied as a vector, each element relating to a different outcome.</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Data frame with harmonised effects and alleles</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Expects data in the format generated by <code><a href="read_exposure_data.html">read_exposure_data()</a></code> and <code><a href="extract_outcome_data.html">extract_outcome_data()</a></code>.
This means the inputs must be dataframes with the following columns:</p>
<p><code>outcome_dat</code>:</p><ul><li><p><code>SNP</code></p></li>
<li><p><code>beta.outcome</code></p></li>
<li><p><code>se.outcome</code></p></li>
<li><p><code>effect_allele.outcome</code></p></li>
<li><p><code>other_allele.outcome</code></p></li>
<li><p><code>eaf.outcome</code></p></li>
<li><p><code>outcome</code></p></li>
</ul><p><code>exposure_dat</code>:</p><ul><li><p><code>SNP</code></p></li>
<li><p><code>beta.exposure</code></p></li>
<li><p><code>se.exposure</code></p></li>
<li><p><code>effect_allele.exposure</code></p></li>
<li><p><code>other_allele.exposure</code></p></li>
<li><p><code>eaf.exposure</code></p></li>
</ul><p>The function tries to harmonise INDELs. If they are coded as sequence strings things work more smoothly.
If they are coded as D/I in one dataset it will try to convert them to sequences if the other dataset has adequate information.
If coded as D/I in one dataset and as a variant with equal length INDEL alleles in the other, the variant is dropped.
If one or both the datasets only has one allele (i.e. the effect allele) then harmonisation is naturally going to be more ambiguous and more variants will be dropped.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

