# Package index

## All functions

- [`Isq()`](https://mrcieu.github.io/TwoSampleMR/reference/Isq.md) :
  I-squared calculation

- [`add_metadata()`](https://mrcieu.github.io/TwoSampleMR/reference/add_metadata.md)
  : Add meta data to extracted data

- [`add_rsq()`](https://mrcieu.github.io/TwoSampleMR/reference/add_rsq.md)
  : Estimate r-squared of each association

- [`allele_frequency()`](https://mrcieu.github.io/TwoSampleMR/reference/allele_frequency.md)
  : Estimate allele frequency from SNP

- [`available_outcomes()`](https://mrcieu.github.io/TwoSampleMR/reference/available_outcomes.md)
  : Get list of studies with available GWAS summary statistics through
  API

- [`clump_data()`](https://mrcieu.github.io/TwoSampleMR/reference/clump_data.md)
  : Perform LD clumping on SNP data

- [`combine_all_mrresults()`](https://mrcieu.github.io/TwoSampleMR/reference/combine_all_mrresults.md)
  : Combine all mr results

- [`combine_data()`](https://mrcieu.github.io/TwoSampleMR/reference/combine_data.md)
  : Combine data

- [`contingency()`](https://mrcieu.github.io/TwoSampleMR/reference/contingency.md)
  : Obtain 2x2 contingency table from marginal parameters and odds ratio

- [`convert_outcome_to_exposure()`](https://mrcieu.github.io/TwoSampleMR/reference/convert_outcome_to_exposure.md)
  : Convert outcome data to exposure data

- [`dat_to_MRInput()`](https://mrcieu.github.io/TwoSampleMR/reference/dat_to_MRInput.md)
  : Convert TwoSampleMR format to MendelianRandomization format

- [`dat_to_RadialMR()`](https://mrcieu.github.io/TwoSampleMR/reference/dat_to_RadialMR.md)
  : Convert dat to RadialMR format

- [`default_parameters()`](https://mrcieu.github.io/TwoSampleMR/reference/default_parameters.md)
  : List of parameters for use with MR functions

- [`directionality_test()`](https://mrcieu.github.io/TwoSampleMR/reference/directionality_test.md)
  : Perform MR Steiger test of directionality

- [`effective_n()`](https://mrcieu.github.io/TwoSampleMR/reference/effective_n.md)
  : Estimate the effective sample size in a case control study

- [`enrichment()`](https://mrcieu.github.io/TwoSampleMR/reference/enrichment.md)
  : Perform enrichment analysis

- [`enrichment_method_list()`](https://mrcieu.github.io/TwoSampleMR/reference/enrichment_method_list.md)
  : Get list of available p-value enrichment methods

- [`estimate_trait_sd()`](https://mrcieu.github.io/TwoSampleMR/reference/estimate_trait_sd.md)
  : Estimate trait SD by obtaining beta estimates from z-scores and
  finding the ratio with original beta values

- [`extract_instruments()`](https://mrcieu.github.io/TwoSampleMR/reference/extract_instruments.md)
  : Find instruments for use in MR from the OpenGWAS database

- [`extract_outcome_data()`](https://mrcieu.github.io/TwoSampleMR/reference/extract_outcome_data.md)
  :

  Supply the output from
  [`read_exposure_data()`](https://mrcieu.github.io/TwoSampleMR/reference/read_exposure_data.md)
  and all the SNPs therein will be queried against the requested
  outcomes in remote database using API.

- [`fishers_combined_test()`](https://mrcieu.github.io/TwoSampleMR/reference/fishers_combined_test.md)
  : Fisher's combined test

- [`forest_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/forest_plot.md)
  : Forest plot for multiple exposures and multiple outcomes

- [`forest_plot_1_to_many()`](https://mrcieu.github.io/TwoSampleMR/reference/forest_plot_1_to_many.md)
  : 1-to-many forest plot

- [`forest_plot_basic2()`](https://mrcieu.github.io/TwoSampleMR/reference/forest_plot_basic2.md)
  : A basic forest plot

- [`format_1_to_many()`](https://mrcieu.github.io/TwoSampleMR/reference/format_1_to_many.md)
  : Format MR results for a 1-to-many forest plot

- [`format_aries_mqtl()`](https://mrcieu.github.io/TwoSampleMR/reference/format_aries_mqtl.md)
  : Get data from methylation QTL results

- [`format_data()`](https://mrcieu.github.io/TwoSampleMR/reference/format_data.md)
  : Read exposure or outcome data

- [`format_gtex_eqtl()`](https://mrcieu.github.io/TwoSampleMR/reference/format_gtex_eqtl.md)
  : Get data from eQTL catalog into correct format

- [`format_gwas_catalog()`](https://mrcieu.github.io/TwoSampleMR/reference/format_gwas_catalog.md)
  : Get data selected from GWAS catalog into correct format

- [`format_metab_qtls()`](https://mrcieu.github.io/TwoSampleMR/reference/format_metab_qtls.md)
  : Get data from metabolomic QTL results

- [`format_mr_results()`](https://mrcieu.github.io/TwoSampleMR/reference/format_mr_results.md)
  : Format MR results for forest plot

- [`format_proteomic_qtls()`](https://mrcieu.github.io/TwoSampleMR/reference/format_proteomic_qtls.md)
  : Get data from proteomic QTL results

- [`generate_odds_ratios()`](https://mrcieu.github.io/TwoSampleMR/reference/generate_odds_ratios.md)
  : Generate odds ratios

- [`get_p_from_r2n()`](https://mrcieu.github.io/TwoSampleMR/reference/get_p_from_r2n.md)
  : Calculate p-value from R-squared and sample size

- [`get_population_allele_frequency()`](https://mrcieu.github.io/TwoSampleMR/reference/get_population_allele_frequency.md)
  : Estimate the allele frequency in population from case/control
  summary data

- [`get_r_from_bsen()`](https://mrcieu.github.io/TwoSampleMR/reference/get_r_from_bsen.md)
  : Estimate R-squared from beta, standard error and sample size

- [`get_r_from_lor()`](https://mrcieu.github.io/TwoSampleMR/reference/get_r_from_lor.md)
  : Estimate proportion of variance of liability explained by SNP in
  general population

- [`get_r_from_pn()`](https://mrcieu.github.io/TwoSampleMR/reference/get_r_from_pn.md)
  : Calculate variance explained from p-values and sample size

- [`get_se()`](https://mrcieu.github.io/TwoSampleMR/reference/get_se.md)
  : Get SE from effect size and p-value

- [`harmonise_data()`](https://mrcieu.github.io/TwoSampleMR/reference/harmonise_data.md)
  : Harmonise the alleles and effects between the exposure and outcome

- [`harmonise_ld_dat()`](https://mrcieu.github.io/TwoSampleMR/reference/harmonise_ld_dat.md)
  : Harmonise LD matrix against summary data

- [`ld_matrix()`](https://mrcieu.github.io/TwoSampleMR/reference/ld_matrix.md)
  : Get LD matrix for list of SNPs

- [`ldsc_h2()`](https://mrcieu.github.io/TwoSampleMR/reference/ldsc_h2.md)
  : Univariate LDSC

- [`ldsc_rg()`](https://mrcieu.github.io/TwoSampleMR/reference/ldsc_rg.md)
  : Bivariate LDSC

- [`make_dat()`](https://mrcieu.github.io/TwoSampleMR/reference/make_dat.md)
  : Convenient function to create a harmonised dataset

- [`mr()`](https://mrcieu.github.io/TwoSampleMR/reference/mr.md) :
  Perform all Mendelian randomization tests

- [`mr_density_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_density_plot.md)
  : Density plot

- [`mr_egger_regression()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_egger_regression.md)
  : Egger's regression for Mendelian randomization

- [`mr_egger_regression_bootstrap()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_egger_regression_bootstrap.md)
  : Run bootstrap to generate standard errors for MR

- [`mr_forest_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_forest_plot.md)
  : Forest plot

- [`mr_funnel_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_funnel_plot.md)
  : Funnel plot

- [`mr_grip()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_grip.md)
  : MR-GRIP: a modified MR-Egger model with the Genotype Recoding
  Invariance Property

- [`mr_heterogeneity()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_heterogeneity.md)
  : Get heterogeneity statistics

- [`mr_ivw()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_ivw.md)
  : Inverse variance weighted regression

- [`mr_ivw_fe()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_ivw_fe.md)
  : Inverse variance weighted regression (fixed effects)

- [`mr_ivw_mre()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_ivw_mre.md)
  : Inverse variance weighted regression (multiplicative random effects
  model)

- [`mr_ivw_radial()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_ivw_radial.md)
  : Radial IVW analysis

- [`mr_leaveoneout()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_leaveoneout.md)
  : Leave one out sensitivity analysis

- [`mr_leaveoneout_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_leaveoneout_plot.md)
  : Plot results from leaveoneout analysis

- [`mr_median()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_median.md)
  : MR median estimators

- [`mr_meta_fixed()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_meta_fixed.md)
  : Perform 2 sample IV using fixed effects meta analysis and delta
  method for standard errors

- [`mr_meta_fixed_simple()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_meta_fixed_simple.md)
  : Perform 2 sample IV using simple standard error

- [`mr_meta_random()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_meta_random.md)
  : Perform 2 sample IV using random effects meta analysis and delta
  method for standard errors

- [`mr_method_list()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_method_list.md)
  : Get list of available MR methods

- [`mr_mode()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_mode.md)
  : MR mode estimators

- [`mr_moe()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_moe.md)
  : Mixture of experts

- [`mr_penalised_weighted_median()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_penalised_weighted_median.md)
  : Penalised weighted median MR

- [`mr_pleiotropy_test()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_pleiotropy_test.md)
  : Test for horizontal pleiotropy in MR analysis

- [`mr_raps()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_raps.md)
  : Robust adjusted profile score

- [`mr_report()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_report.md)
  : Generate MR report

- [`mr_rucker()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_rucker.md)
  : MR Rucker framework

- [`mr_rucker_bootstrap()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_rucker_bootstrap.md)
  : Run rucker with bootstrap estimates

- [`mr_rucker_cooksdistance()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_rucker_cooksdistance.md)
  : MR Rucker with outliers automatically detected and removed

- [`mr_rucker_jackknife()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_rucker_jackknife.md)
  : Run rucker with jackknife estimates

- [`mr_scatter_plot()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_scatter_plot.md)
  : Create scatter plot with fitted lines showing the causal effect
  estimate for different MR estimators

- [`mr_sign()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_sign.md)
  : MR sign test

- [`mr_simple_median()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_simple_median.md)
  : Simple median method

- [`mr_simple_mode()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_simple_mode.md)
  : MR simple mode estimator

- [`mr_simple_mode_nome()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_simple_mode_nome.md)
  : MR simple mode estimator (NOME)

- [`mr_singlesnp()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_singlesnp.md)
  : Perform 2 sample MR on each SNP individually

- [`mr_steiger()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_steiger.md)
  : MR Steiger test of directionality

- [`mr_steiger2()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_steiger2.md)
  : MR Steiger test of directionality

- [`mr_two_sample_ml()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_two_sample_ml.md)
  : Maximum likelihood MR method

- [`mr_uwr()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_uwr.md)
  : Unweighted regression

- [`mr_wald_ratio()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_wald_ratio.md)
  : Perform 2 sample IV using Wald ratio.

- [`mr_weighted_median()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_weighted_median.md)
  : Weighted median method

- [`mr_weighted_mode()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_weighted_mode.md)
  : MR weighted mode estimator

- [`mr_weighted_mode_nome()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_weighted_mode_nome.md)
  : MR weighted mode estimator (NOME)

- [`mr_wrapper()`](https://mrcieu.github.io/TwoSampleMR/reference/mr_wrapper.md)
  : Perform full set of MR analyses

- [`mv_basic()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_basic.md)
  : Perform basic multivariable MR

- [`mv_extract_exposures()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_extract_exposures.md)
  : Extract exposure variables for multivariable MR

- [`mv_extract_exposures_local()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_extract_exposures_local.md)
  : Attempt to perform MVMR using local data

- [`mv_harmonise_data()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_harmonise_data.md)
  : Harmonise exposure and outcome for multivariable MR

- [`mv_ivw()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_ivw.md)
  : Perform IVW multivariable MR

- [`mv_lasso_feature_selection()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_lasso_feature_selection.md)
  : Apply LASSO feature selection to mvdat object

- [`mv_multiple()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_multiple.md)
  : Perform IVW multivariable MR

- [`mv_residual()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_residual.md)
  : Perform basic multivariable MR

- [`mv_subset()`](https://mrcieu.github.io/TwoSampleMR/reference/mv_subset.md)
  : Perform multivariable MR on subset of features

- [`power_prune()`](https://mrcieu.github.io/TwoSampleMR/reference/power_prune.md)
  : Power prune

- [`read_exposure_data()`](https://mrcieu.github.io/TwoSampleMR/reference/read_exposure_data.md)
  : Read exposure data

- [`read_outcome_data()`](https://mrcieu.github.io/TwoSampleMR/reference/read_outcome_data.md)
  : Read outcome data

- [`run_mr_presso()`](https://mrcieu.github.io/TwoSampleMR/reference/run_mr_presso.md)
  : Wrapper for MR-PRESSO

- [`run_mrmix()`](https://mrcieu.github.io/TwoSampleMR/reference/run_mrmix.md)
  : Perform MRMix analysis on harmonised dat object

- [`size.prune()`](https://mrcieu.github.io/TwoSampleMR/reference/size.prune.md)
  : Size prune

- [`sort_1_to_many()`](https://mrcieu.github.io/TwoSampleMR/reference/sort_1_to_many.md)
  : Sort results for 1-to-many forest plot

- [`split_exposure()`](https://mrcieu.github.io/TwoSampleMR/reference/split_exposure.md)
  : Split exposure column

- [`split_outcome()`](https://mrcieu.github.io/TwoSampleMR/reference/split_outcome.md)
  : Split outcome column

- [`standardise_units()`](https://mrcieu.github.io/TwoSampleMR/reference/standardise_units.md)
  : Try to standardise continuous traits to be in standard deviation
  units

- [`steiger_filtering()`](https://mrcieu.github.io/TwoSampleMR/reference/steiger_filtering.md)
  : Steiger filtering function

- [`steiger_sensitivity()`](https://mrcieu.github.io/TwoSampleMR/reference/steiger_sensitivity.md)
  : Evaluate the Steiger test's sensitivity to measurement error

- [`subset_on_method()`](https://mrcieu.github.io/TwoSampleMR/reference/subset_on_method.md)
  : Subset MR-results on method

- [`trim()`](https://mrcieu.github.io/TwoSampleMR/reference/trim.md) :
  Trim function to remove leading and trailing blank spaces

- [`weighted_median()`](https://mrcieu.github.io/TwoSampleMR/reference/weighted_median.md)
  : Weighted median method

- [`weighted_median_bootstrap()`](https://mrcieu.github.io/TwoSampleMR/reference/weighted_median_bootstrap.md)
  : Calculate standard errors for weighted median method using bootstrap
