<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ TwoSampleMR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide"><li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="Isq.html">Isq()</a></code>

        </dt>
        <dd>I-squared calculation</dd>
      </dl><dl><dt>

          <code><a href="add_metadata.html">add_metadata()</a></code>

        </dt>
        <dd>Add meta data to extracted data</dd>
      </dl><dl><dt>

          <code><a href="add_rsq.html">add_rsq()</a></code>

        </dt>
        <dd>Estimate r-square of each association</dd>
      </dl><dl><dt>

          <code><a href="allele_frequency.html">allele_frequency()</a></code>

        </dt>
        <dd>Estimate allele frequency from SNP</dd>
      </dl><dl><dt>

          <code><a href="available_outcomes.html">available_outcomes()</a></code>

        </dt>
        <dd>Get list of studies with available GWAS summary statistics through API</dd>
      </dl><dl><dt>

          <code><a href="clump_data.html">clump_data()</a></code>

        </dt>
        <dd>Perform LD clumping on SNP data</dd>
      </dl><dl><dt>

          <code><a href="combine_all_mrresults.html">combine_all_mrresults()</a></code>

        </dt>
        <dd>Combine all mr results</dd>
      </dl><dl><dt>

          <code><a href="combine_data.html">combine_data()</a></code>

        </dt>
        <dd>Combine data</dd>
      </dl><dl><dt>

          <code><a href="contingency.html">contingency()</a></code>

        </dt>
        <dd>Obtain 2x2 contingency table from marginal parameters and odds ratio</dd>
      </dl><dl><dt>

          <code><a href="convert_outcome_to_exposure.html">convert_outcome_to_exposure()</a></code>

        </dt>
        <dd>Convert outcome data to exposure data</dd>
      </dl><dl><dt>

          <code><a href="dat_to_MRInput.html">dat_to_MRInput()</a></code>

        </dt>
        <dd>Convert TwoSampleMR format to MendelianRandomization format</dd>
      </dl><dl><dt>

          <code><a href="dat_to_RadialMR.html">dat_to_RadialMR()</a></code>

        </dt>
        <dd>Convert dat to RadialMR format</dd>
      </dl><dl><dt>

          <code><a href="default_parameters.html">default_parameters()</a></code>

        </dt>
        <dd>List of parameters for use with MR functions</dd>
      </dl><dl><dt>

          <code><a href="directionality_test.html">directionality_test()</a></code>

        </dt>
        <dd>Perform MR Steiger test of directionality</dd>
      </dl><dl><dt>

          <code><a href="effective_n.html">effective_n()</a></code>

        </dt>
        <dd>Estimate the effective sample size in a case control study</dd>
      </dl><dl><dt>

          <code><a href="enrichment.html">enrichment()</a></code>

        </dt>
        <dd>Perform enrichment analysis</dd>
      </dl><dl><dt>

          <code><a href="enrichment_method_list.html">enrichment_method_list()</a></code>

        </dt>
        <dd>Get list of available p-value enrichment methods</dd>
      </dl><dl><dt>

          <code><a href="estimate_trait_sd.html">estimate_trait_sd()</a></code>

        </dt>
        <dd>Estimate trait SD by obtaining beta estimates from z-scores and finding the ratio with original beta values</dd>
      </dl><dl><dt>

          <code><a href="extract_instruments.html">extract_instruments()</a></code>

        </dt>
        <dd>Find instruments for use in MR from the MR Base database</dd>
      </dl><dl><dt>

          <code><a href="extract_outcome_data.html">extract_outcome_data()</a></code>

        </dt>
        <dd>Supply the output from <code><a href="../reference/read_exposure_data.html">read_exposure_data()</a></code> and all the SNPs therein will be queried against the requested outcomes in remote database using API.</dd>
      </dl><dl><dt>

          <code><a href="fishers_combined_test.html">fishers_combined_test()</a></code>

        </dt>
        <dd>Fisher's combined test</dd>
      </dl><dl><dt>

          <code><a href="forest_plot.html">forest_plot()</a></code>

        </dt>
        <dd>Forest plot for multiple exposures and multiple outcomes</dd>
      </dl><dl><dt>

          <code><a href="forest_plot_1_to_many.html">forest_plot_1_to_many()</a></code>

        </dt>
        <dd>1-to-many forest plot</dd>
      </dl><dl><dt>

          <code><a href="forest_plot_basic2.html">forest_plot_basic2()</a></code>

        </dt>
        <dd>A basic forest plot</dd>
      </dl><dl><dt>

          <code><a href="format_1_to_many.html">format_1_to_many()</a></code>

        </dt>
        <dd>Format MR results for a 1-to-many forest plot</dd>
      </dl><dl><dt>

          <code><a href="format_aries_mqtl.html">format_aries_mqtl()</a></code>

        </dt>
        <dd>Get data from methylation QTL results</dd>
      </dl><dl><dt>

          <code><a href="format_data.html">format_data()</a></code>

        </dt>
        <dd>Read exposure or outcome data</dd>
      </dl><dl><dt>

          <code><a href="format_gtex_eqtl.html">format_gtex_eqtl()</a></code>

        </dt>
        <dd>Get data from eQTL catalog into correct format</dd>
      </dl><dl><dt>

          <code><a href="format_gwas_catalog.html">format_gwas_catalog()</a></code>

        </dt>
        <dd>Get data selected from GWAS catalog into correct format</dd>
      </dl><dl><dt>

          <code><a href="format_metab_qtls.html">format_metab_qtls()</a></code>

        </dt>
        <dd>Get data from metabolomic QTL results</dd>
      </dl><dl><dt>

          <code><a href="format_mr_results.html">format_mr_results()</a></code>

        </dt>
        <dd>Format MR results for forest plot</dd>
      </dl><dl><dt>

          <code><a href="format_proteomic_qtls.html">format_proteomic_qtls()</a></code>

        </dt>
        <dd>Get data from proteomic QTL results</dd>
      </dl><dl><dt>

          <code><a href="generate_odds_ratios.html">generate_odds_ratios()</a></code>

        </dt>
        <dd>Generate odds ratios</dd>
      </dl><dl><dt>

          <code><a href="get_p_from_r2n.html">get_p_from_r2n()</a></code>

        </dt>
        <dd>Calculate p-value from R-squared and sample size</dd>
      </dl><dl><dt>

          <code><a href="get_population_allele_frequency.html">get_population_allele_frequency()</a></code>

        </dt>
        <dd>Estimate the allele frequency in population from case/control summary data</dd>
      </dl><dl><dt>

          <code><a href="get_r_from_bsen.html">get_r_from_bsen()</a></code>

        </dt>
        <dd>Estimate R-squared from beta, standard error and sample size</dd>
      </dl><dl><dt>

          <code><a href="get_r_from_lor.html">get_r_from_lor()</a></code>

        </dt>
        <dd>Estimate proportion of variance of liability explained by SNP in general population</dd>
      </dl><dl><dt>

          <code><a href="get_r_from_pn.html">get_r_from_pn()</a></code>

        </dt>
        <dd>Calculate variance explained from p-values and sample size</dd>
      </dl><dl><dt>

          <code><a href="get_se.html">get_se()</a></code>

        </dt>
        <dd>Get SE from effect size and p-value</dd>
      </dl><dl><dt>

          <code><a href="harmonise_data.html">harmonise_data()</a></code>

        </dt>
        <dd>Harmonise the alleles and effects between the exposure and outcome</dd>
      </dl><dl><dt>

          <code><a href="harmonise_ld_dat.html">harmonise_ld_dat()</a></code>

        </dt>
        <dd>Harmonise LD matrix against summary data</dd>
      </dl><dl><dt>

          <code><a href="ld_matrix.html">ld_matrix()</a></code>

        </dt>
        <dd>Get LD matrix for list of SNPs</dd>
      </dl><dl><dt>

          <code><a href="ldsc_h2.html">ldsc_h2()</a></code>

        </dt>
        <dd>Univariate LDSC</dd>
      </dl><dl><dt>

          <code><a href="ldsc_rg.html">ldsc_rg()</a></code>

        </dt>
        <dd>Bivariate LDSC</dd>
      </dl><dl><dt>

          <code><a href="make_dat.html">make_dat()</a></code>

        </dt>
        <dd>Convenient function to create a harmonised dataset</dd>
      </dl><dl><dt>

          <code><a href="mr.html">mr()</a></code>

        </dt>
        <dd>Perform all Mendelian randomization tests</dd>
      </dl><dl><dt>

          <code><a href="mr_density_plot.html">mr_density_plot()</a></code>

        </dt>
        <dd>Density plot</dd>
      </dl><dl><dt>

          <code><a href="mr_egger_regression.html">mr_egger_regression()</a></code>

        </dt>
        <dd>Egger's regression for Mendelian randomization</dd>
      </dl><dl><dt>

          <code><a href="mr_egger_regression_bootstrap.html">mr_egger_regression_bootstrap()</a></code>

        </dt>
        <dd>Run bootstrap to generate standard errors for MR</dd>
      </dl><dl><dt>

          <code><a href="mr_forest_plot.html">mr_forest_plot()</a></code>

        </dt>
        <dd>Forest plot</dd>
      </dl><dl><dt>

          <code><a href="mr_funnel_plot.html">mr_funnel_plot()</a></code>

        </dt>
        <dd>Funnel plot</dd>
      </dl><dl><dt>

          <code><a href="mr_heterogeneity.html">mr_heterogeneity()</a></code>

        </dt>
        <dd>Get heterogeneity statistics</dd>
      </dl><dl><dt>

          <code><a href="mr_ivw.html">mr_ivw()</a></code>

        </dt>
        <dd>Inverse variance weighted regression</dd>
      </dl><dl><dt>

          <code><a href="mr_ivw_fe.html">mr_ivw_fe()</a></code>

        </dt>
        <dd>Inverse variance weighted regression (fixed effects)</dd>
      </dl><dl><dt>

          <code><a href="mr_ivw_mre.html">mr_ivw_mre()</a></code>

        </dt>
        <dd>Inverse variance weighted regression (multiplicative random effects model)</dd>
      </dl><dl><dt>

          <code><a href="mr_ivw_radial.html">mr_ivw_radial()</a></code>

        </dt>
        <dd>Radial IVW analysis</dd>
      </dl><dl><dt>

          <code><a href="mr_leaveoneout.html">mr_leaveoneout()</a></code>

        </dt>
        <dd>Leave one out sensitivity analysis</dd>
      </dl><dl><dt>

          <code><a href="mr_leaveoneout_plot.html">mr_leaveoneout_plot()</a></code>

        </dt>
        <dd>Plot results from leaveoneout analysis</dd>
      </dl><dl><dt>

          <code><a href="mr_median.html">mr_median()</a></code>

        </dt>
        <dd>MR median estimators</dd>
      </dl><dl><dt>

          <code><a href="mr_meta_fixed.html">mr_meta_fixed()</a></code>

        </dt>
        <dd>Perform 2 sample IV using fixed effects meta analysis and delta method for standard errors</dd>
      </dl><dl><dt>

          <code><a href="mr_meta_fixed_simple.html">mr_meta_fixed_simple()</a></code>

        </dt>
        <dd>Perform 2 sample IV using simple standard error</dd>
      </dl><dl><dt>

          <code><a href="mr_meta_random.html">mr_meta_random()</a></code>

        </dt>
        <dd>Perform 2 sample IV using random effects meta analysis and delta method for standard errors</dd>
      </dl><dl><dt>

          <code><a href="mr_method_list.html">mr_method_list()</a></code>

        </dt>
        <dd>Get list of available MR methods</dd>
      </dl><dl><dt>

          <code><a href="mr_mode.html">mr_mode()</a></code>

        </dt>
        <dd>MR mode estimators</dd>
      </dl><dl><dt>

          <code><a href="mr_moe.html">mr_moe()</a></code>

        </dt>
        <dd>Mixture of experts</dd>
      </dl><dl><dt>

          <code><a href="mr_penalised_weighted_median.html">mr_penalised_weighted_median()</a></code>

        </dt>
        <dd>Penalised weighted median MR</dd>
      </dl><dl><dt>

          <code><a href="mr_pleiotropy_test.html">mr_pleiotropy_test()</a></code>

        </dt>
        <dd>Test for horizontal pleiotropy in MR analysis</dd>
      </dl><dl><dt>

          <code><a href="mr_raps.html">mr_raps()</a></code>

        </dt>
        <dd>Robust adjusted profile score</dd>
      </dl><dl><dt>

          <code><a href="mr_report.html">mr_report()</a></code>

        </dt>
        <dd>Generate MR report</dd>
      </dl><dl><dt>

          <code><a href="mr_rucker.html">mr_rucker()</a></code>

        </dt>
        <dd>MR Rucker framework</dd>
      </dl><dl><dt>

          <code><a href="mr_rucker_bootstrap.html">mr_rucker_bootstrap()</a></code>

        </dt>
        <dd>Run rucker with bootstrap estimates</dd>
      </dl><dl><dt>

          <code><a href="mr_rucker_cooksdistance.html">mr_rucker_cooksdistance()</a></code>

        </dt>
        <dd>MR Rucker with outliers automatically detected and removed</dd>
      </dl><dl><dt>

          <code><a href="mr_rucker_jackknife.html">mr_rucker_jackknife()</a></code>

        </dt>
        <dd>Run rucker with jackknife estimates</dd>
      </dl><dl><dt>

          <code><a href="mr_scatter_plot.html">mr_scatter_plot()</a></code>

        </dt>
        <dd>Create scatter plot with lines showing the causal estimate for different MR tests</dd>
      </dl><dl><dt>

          <code><a href="mr_sign.html">mr_sign()</a></code>

        </dt>
        <dd>MR sign test</dd>
      </dl><dl><dt>

          <code><a href="mr_simple_median.html">mr_simple_median()</a></code>

        </dt>
        <dd>Simple median method</dd>
      </dl><dl><dt>

          <code><a href="mr_simple_mode.html">mr_simple_mode()</a></code>

        </dt>
        <dd>MR simple mode estimator</dd>
      </dl><dl><dt>

          <code><a href="mr_simple_mode_nome.html">mr_simple_mode_nome()</a></code>

        </dt>
        <dd>MR simple mode estimator (NOME)</dd>
      </dl><dl><dt>

          <code><a href="mr_singlesnp.html">mr_singlesnp()</a></code>

        </dt>
        <dd>Perform 2 sample MR on each SNP individually</dd>
      </dl><dl><dt>

          <code><a href="mr_steiger.html">mr_steiger()</a></code>

        </dt>
        <dd>MR Steiger test of directionality</dd>
      </dl><dl><dt>

          <code><a href="mr_steiger2.html">mr_steiger2()</a></code>

        </dt>
        <dd>MR Steiger test of directionality</dd>
      </dl><dl><dt>

          <code><a href="mr_two_sample_ml.html">mr_two_sample_ml()</a></code>

        </dt>
        <dd>Maximum likelihood MR method</dd>
      </dl><dl><dt>

          <code><a href="mr_uwr.html">mr_uwr()</a></code>

        </dt>
        <dd>Unweighted regression</dd>
      </dl><dl><dt>

          <code><a href="mr_wald_ratio.html">mr_wald_ratio()</a></code>

        </dt>
        <dd>Perform 2 sample IV using Wald ratio.</dd>
      </dl><dl><dt>

          <code><a href="mr_weighted_median.html">mr_weighted_median()</a></code>

        </dt>
        <dd>Weighted median method</dd>
      </dl><dl><dt>

          <code><a href="mr_weighted_mode.html">mr_weighted_mode()</a></code>

        </dt>
        <dd>MR weighted mode estimator</dd>
      </dl><dl><dt>

          <code><a href="mr_weighted_mode_nome.html">mr_weighted_mode_nome()</a></code>

        </dt>
        <dd>MR weighted mode estimator (NOME)</dd>
      </dl><dl><dt>

          <code><a href="mr_wrapper.html">mr_wrapper()</a></code>

        </dt>
        <dd>Perform full set of MR analyses</dd>
      </dl><dl><dt>

          <code><a href="mv_basic.html">mv_basic()</a></code>

        </dt>
        <dd>Perform basic multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_extract_exposures.html">mv_extract_exposures()</a></code>

        </dt>
        <dd>Extract exposure variables for multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_extract_exposures_local.html">mv_extract_exposures_local()</a></code>

        </dt>
        <dd>Attempt to perform MVMR using local data</dd>
      </dl><dl><dt>

          <code><a href="mv_harmonise_data.html">mv_harmonise_data()</a></code>

        </dt>
        <dd>Harmonise exposure and outcome for multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_ivw.html">mv_ivw()</a></code>

        </dt>
        <dd>Perform IVW multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_lasso_feature_selection.html">mv_lasso_feature_selection()</a></code>

        </dt>
        <dd>Apply LASSO feature selection to mvdat object</dd>
      </dl><dl><dt>

          <code><a href="mv_multiple.html">mv_multiple()</a></code>

        </dt>
        <dd>Perform IVW multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_residual.html">mv_residual()</a></code>

        </dt>
        <dd>Perform basic multivariable MR</dd>
      </dl><dl><dt>

          <code><a href="mv_subset.html">mv_subset()</a></code>

        </dt>
        <dd>Perform multivariable MR on subset of features</dd>
      </dl><dl><dt>

          <code><a href="power_prune.html">power_prune()</a></code>

        </dt>
        <dd>Power prune</dd>
      </dl><dl><dt>

          <code><a href="read_exposure_data.html">read_exposure_data()</a></code>

        </dt>
        <dd>Read exposure data</dd>
      </dl><dl><dt>

          <code><a href="read_outcome_data.html">read_outcome_data()</a></code>

        </dt>
        <dd>Read outcome data</dd>
      </dl><dl><dt>

          <code><a href="run_mr_presso.html">run_mr_presso()</a></code>

        </dt>
        <dd>Wrapper for MR-PRESSO</dd>
      </dl><dl><dt>

          <code><a href="run_mrmix.html">run_mrmix()</a></code>

        </dt>
        <dd>Perform MRMix analysis on harmonised dat object</dd>
      </dl><dl><dt>

          <code><a href="size.prune.html">size.prune()</a></code>

        </dt>
        <dd>Size prune</dd>
      </dl><dl><dt>

          <code><a href="sort_1_to_many.html">sort_1_to_many()</a></code>

        </dt>
        <dd>Sort results for 1-to-many forest plot</dd>
      </dl><dl><dt>

          <code><a href="split_exposure.html">split_exposure()</a></code>

        </dt>
        <dd>Split exposure column</dd>
      </dl><dl><dt>

          <code><a href="split_outcome.html">split_outcome()</a></code>

        </dt>
        <dd>Split outcome column</dd>
      </dl><dl><dt>

          <code><a href="standardise_units.html">standardise_units()</a></code>

        </dt>
        <dd>Try to standardise continuous traits to be in standard deviation units</dd>
      </dl><dl><dt>

          <code><a href="steiger_filtering.html">steiger_filtering()</a></code>

        </dt>
        <dd>Steiger filtering function</dd>
      </dl><dl><dt>

          <code><a href="steiger_sensitivity.html">steiger_sensitivity()</a></code>

        </dt>
        <dd>Evaluate the Steiger test's sensitivity to measurement error</dd>
      </dl><dl><dt>

          <code><a href="subset_on_method.html">subset_on_method()</a></code>

        </dt>
        <dd>Subset MR-results on method</dd>
      </dl><dl><dt>

          <code><a href="trim.html">trim()</a></code>

        </dt>
        <dd>Trim function to remove leading and trailing blank spaces</dd>
      </dl><dl><dt>

          <code><a href="weighted_median.html">weighted_median()</a></code>

        </dt>
        <dd>Weighted median method</dd>
      </dl><dl><dt>

          <code><a href="weighted_median_bootstrap.html">weighted_median_bootstrap()</a></code>

        </dt>
        <dd>Calculate standard errors for weighted median method using bootstrap</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

