<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get LD matrix for list of SNPs — ld_matrix • TwoSampleMR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Get LD matrix for list of SNPs — ld_matrix"><meta name="description" content="This function takes a list of SNPs and searches for them in a specified super-population in the 1000 Genomes phase 3 reference panel.
It then creates an LD matrix of r values (signed, and not squared).
All LD values are with respect to the major alleles in the 1000G dataset.
You can specify whether the allele names are displayed."><meta property="og:description" content="This function takes a list of SNPs and searches for them in a specified super-population in the 1000 Genomes phase 3 reference panel.
It then creates an LD matrix of r values (signed, and not squared).
All LD values are with respect to the major alleles in the 1000G dataset.
You can specify whether the allele names are displayed."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.18</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide"><li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get LD matrix for list of SNPs</h1>
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/v0.6.18/R/ld.R" class="external-link"><code>R/ld.R</code></a></small>
      <div class="d-none name"><code>ld_matrix.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes a list of SNPs and searches for them in a specified super-population in the 1000 Genomes phase 3 reference panel.
It then creates an LD matrix of r values (signed, and not squared).
All LD values are with respect to the major alleles in the 1000G dataset.
You can specify whether the allele names are displayed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ld_matrix</span><span class="op">(</span><span class="va">snps</span>, with_alleles <span class="op">=</span> <span class="cn">TRUE</span>, pop <span class="op">=</span> <span class="st">"EUR"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-snps">snps<a class="anchor" aria-label="anchor" href="#arg-snps"></a></dt>
<dd><p>List of SNPs.</p></dd>


<dt id="arg-with-alleles">with_alleles<a class="anchor" aria-label="anchor" href="#arg-with-alleles"></a></dt>
<dd><p>Whether to append the allele names to the SNP names. The default is <code>TRUE</code>.</p></dd>


<dt id="arg-pop">pop<a class="anchor" aria-label="anchor" href="#arg-pop"></a></dt>
<dd><p>Super-population to use as reference panel. Default = <code>"EUR"</code>. Options are <code>"EUR"</code>, <code>"SAS"</code>, <code>"EAS"</code>, <code>"AFR"</code>, <code>"AMR"</code>. <code>'legacy'</code> also available - which is a previously used version of the EUR panel with a slightly different set of markers.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Matrix of LD r values</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The data used for generating the LD matrix includes only bi-allelic SNPs with MAF &gt; 0.01,
so it's quite possible that a variant you want to include will be absent.
If it is absent, it will be automatically excluded from the results.</p>
<p>You can check if your variants are present in the LD reference panel using <code><a href="https://mrcieu.github.io/ieugwasr/reference/ld_reflookup.html" class="external-link">ieugwasr::ld_reflookup()</a></code>.</p>
<p>This function does put load on the OpenGWAS servers, which makes life more difficult for other users,
and has been limited to analyse only up to 500 variants at a time.
We have implemented a method and made available the LD reference panels to perform the operation locally,
see <code><a href="https://mrcieu.github.io/ieugwasr/reference/ld_matrix.html" class="external-link">ieugwasr::ld_matrix()</a></code> and related vignettes for details.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

