<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Attempt to perform MVMR using local data — mv_extract_exposures_local • TwoSampleMR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Attempt to perform MVMR using local data — mv_extract_exposures_local"><meta name="description" content="Allows you to read in summary data from text files to format the multivariable exposure dataset."><meta property="og:description" content="Allows you to read in summary data from text files to format the multivariable exposure dataset."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.23</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide"><li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Attempt to perform MVMR using local data</h1>
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/R/multivariable_mr.R" class="external-link"><code>R/multivariable_mr.R</code></a></small>
      <div class="d-none name"><code>mv_extract_exposures_local.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Allows you to read in summary data from text files to format the multivariable exposure dataset.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mv_extract_exposures_local</span><span class="op">(</span></span>
<span>  <span class="va">filenames_exposure</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">" "</span>,</span>
<span>  phenotype_col <span class="op">=</span> <span class="st">"Phenotype"</span>,</span>
<span>  snp_col <span class="op">=</span> <span class="st">"SNP"</span>,</span>
<span>  beta_col <span class="op">=</span> <span class="st">"beta"</span>,</span>
<span>  se_col <span class="op">=</span> <span class="st">"se"</span>,</span>
<span>  eaf_col <span class="op">=</span> <span class="st">"eaf"</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>  other_allele_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>  pval_col <span class="op">=</span> <span class="st">"pval"</span>,</span>
<span>  units_col <span class="op">=</span> <span class="st">"units"</span>,</span>
<span>  ncase_col <span class="op">=</span> <span class="st">"ncase"</span>,</span>
<span>  ncontrol_col <span class="op">=</span> <span class="st">"ncontrol"</span>,</span>
<span>  samplesize_col <span class="op">=</span> <span class="st">"samplesize"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  id_col <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>  min_pval <span class="op">=</span> <span class="fl">1e-200</span>,</span>
<span>  log_pval <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pval_threshold <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clump_r2 <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  clump_kb <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  pop <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  harmonise_strictness <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-filenames-exposure">filenames_exposure<a class="anchor" aria-label="anchor" href="#arg-filenames-exposure"></a></dt>
<dd><p>Filenames for each exposure dataset. Must have header with at least SNP column present. Following arguments are used for determining how to read the filename and clumping etc.</p></dd>


<dt id="arg-sep">sep<a class="anchor" aria-label="anchor" href="#arg-sep"></a></dt>
<dd><p>Specify delimiter in file. The default is space, i.e. <code>sep=" "</code>. If length is 1 it will use the same <code>sep</code> value for each exposure dataset. You can provide a vector of values, one for each exposure dataset, if the values are different across datasets. The same applies to all dataset-formatting options listed below.</p></dd>


<dt id="arg-phenotype-col">phenotype_col<a class="anchor" aria-label="anchor" href="#arg-phenotype-col"></a></dt>
<dd><p>Optional column name for the column with phenotype name corresponding the the SNP. If not present then will be created with the value <code>"Outcome"</code>. Default is <code>"Phenotype"</code>.</p></dd>


<dt id="arg-snp-col">snp_col<a class="anchor" aria-label="anchor" href="#arg-snp-col"></a></dt>
<dd><p>Required name of column with SNP rs IDs. The default is <code>"SNP"</code>.</p></dd>


<dt id="arg-beta-col">beta_col<a class="anchor" aria-label="anchor" href="#arg-beta-col"></a></dt>
<dd><p>Required for MR. Name of column with effect sizes. The default is <code>"beta"</code>.</p></dd>


<dt id="arg-se-col">se_col<a class="anchor" aria-label="anchor" href="#arg-se-col"></a></dt>
<dd><p>Required for MR. Name of column with standard errors. The default is <code>"se"</code>.</p></dd>


<dt id="arg-eaf-col">eaf_col<a class="anchor" aria-label="anchor" href="#arg-eaf-col"></a></dt>
<dd><p>Required for MR. Name of column with effect allele frequency. The default is <code>"eaf"</code>.</p></dd>


<dt id="arg-effect-allele-col">effect_allele_col<a class="anchor" aria-label="anchor" href="#arg-effect-allele-col"></a></dt>
<dd><p>Required for MR. Name of column with effect allele. Must be "A", "C", "T" or "G". The default is <code>"effect_allele"</code>.</p></dd>


<dt id="arg-other-allele-col">other_allele_col<a class="anchor" aria-label="anchor" href="#arg-other-allele-col"></a></dt>
<dd><p>Required for MR. Name of column with non effect allele. Must be "A", "C", "T" or "G". The default is <code>"other_allele"</code>.</p></dd>


<dt id="arg-pval-col">pval_col<a class="anchor" aria-label="anchor" href="#arg-pval-col"></a></dt>
<dd><p>Required for enrichment tests. Name of column with p-value. The default is <code>"pval"</code>.</p></dd>


<dt id="arg-units-col">units_col<a class="anchor" aria-label="anchor" href="#arg-units-col"></a></dt>
<dd><p>Optional column name for units. The default is <code>"units"</code>.</p></dd>


<dt id="arg-ncase-col">ncase_col<a class="anchor" aria-label="anchor" href="#arg-ncase-col"></a></dt>
<dd><p>Optional column name for number of cases. The default is <code>"ncase"</code>.</p></dd>


<dt id="arg-ncontrol-col">ncontrol_col<a class="anchor" aria-label="anchor" href="#arg-ncontrol-col"></a></dt>
<dd><p>Optional column name for number of controls. The default is <code>"ncontrol"</code>.</p></dd>


<dt id="arg-samplesize-col">samplesize_col<a class="anchor" aria-label="anchor" href="#arg-samplesize-col"></a></dt>
<dd><p>Optional column name for sample size. The default is <code>"samplesize"</code>.</p></dd>


<dt id="arg-gene-col">gene_col<a class="anchor" aria-label="anchor" href="#arg-gene-col"></a></dt>
<dd><p>Optional column name for gene name. The default is <code>"gene"</code>.</p></dd>


<dt id="arg-id-col">id_col<a class="anchor" aria-label="anchor" href="#arg-id-col"></a></dt>
<dd><p>Optional column name to give the dataset an ID. Will be generated automatically if not provided for every trait / unit combination. The default is <code>"id"</code>.</p></dd>


<dt id="arg-min-pval">min_pval<a class="anchor" aria-label="anchor" href="#arg-min-pval"></a></dt>
<dd><p>Minimum allowed p-value. The default is <code>1e-200</code>.</p></dd>


<dt id="arg-log-pval">log_pval<a class="anchor" aria-label="anchor" href="#arg-log-pval"></a></dt>
<dd><p>The pval is -log10(P). The default is <code>FALSE</code>.</p></dd>


<dt id="arg-pval-threshold">pval_threshold<a class="anchor" aria-label="anchor" href="#arg-pval-threshold"></a></dt>
<dd><p>Default=<code>5e-8</code> for clumping</p></dd>


<dt id="arg-plink-bin">plink_bin<a class="anchor" aria-label="anchor" href="#arg-plink-bin"></a></dt>
<dd><p>If <code>NULL</code> and <code>bfile</code> is not <code>NULL</code> then will detect packaged plink binary for specific OS. Otherwise specify path to plink binary. Default = <code>NULL</code></p></dd>


<dt id="arg-bfile">bfile<a class="anchor" aria-label="anchor" href="#arg-bfile"></a></dt>
<dd><p>If this is provided then will use the API. Default = <code>NULL</code></p></dd>


<dt id="arg-clump-r-">clump_r2<a class="anchor" aria-label="anchor" href="#arg-clump-r-"></a></dt>
<dd><p>Default=<code>0.001</code> for clumping</p></dd>


<dt id="arg-clump-kb">clump_kb<a class="anchor" aria-label="anchor" href="#arg-clump-kb"></a></dt>
<dd><p>Default=<code>10000</code> for clumping</p></dd>


<dt id="arg-pop">pop<a class="anchor" aria-label="anchor" href="#arg-pop"></a></dt>
<dd><p>Which 1000 genomes super population to use for clumping when using the server</p></dd>


<dt id="arg-harmonise-strictness">harmonise_strictness<a class="anchor" aria-label="anchor" href="#arg-harmonise-strictness"></a></dt>
<dd><p>See action argument in <code><a href="harmonise_data.html">harmonise_data()</a></code>. Default=<code>2</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>List</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that you can provide an array of column names for each column, which is of length <code>filenames_exposure</code></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

