<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Extract exposure variables for multivariable MR — mv_extract_exposures • TwoSampleMR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Extract exposure variables for multivariable MR — mv_extract_exposures"><meta name="description" content="Requires a list of IDs from available_outcomes. For each ID, it extracts instruments. Then, it gets the full list of all instruments and extracts those SNPs for every exposure. Finally, it keeps only the SNPs that are a) independent and b) present in all exposures, and harmonises them to be all on the same strand."><meta property="og:description" content="Requires a list of IDs from available_outcomes. For each ID, it extracts instruments. Then, it gets the full list of all instruments and extracts those SNPs for every exposure. Finally, it keeps only the SNPs that are a) independent and b) present in all exposures, and harmonises them to be all on the same strand."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.24</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide"><li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Extract exposure variables for multivariable MR</h1>
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/R/multivariable_mr.R" class="external-link"><code>R/multivariable_mr.R</code></a></small>
      <div class="d-none name"><code>mv_extract_exposures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Requires a list of IDs from available_outcomes. For each ID, it extracts instruments. Then, it gets the full list of all instruments and extracts those SNPs for every exposure. Finally, it keeps only the SNPs that are a) independent and b) present in all exposures, and harmonises them to be all on the same strand.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mv_extract_exposures</span><span class="op">(</span></span>
<span>  <span class="va">id_exposure</span>,</span>
<span>  clump_r2 <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  clump_kb <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  harmonise_strictness <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  opengwas_jwt <span class="op">=</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://mrcieu.github.io/ieugwasr/reference/get_opengwas_jwt.html" class="external-link">get_opengwas_jwt</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  find_proxies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  force_server <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pval_threshold <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  pop <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-id-exposure">id_exposure<a class="anchor" aria-label="anchor" href="#arg-id-exposure"></a></dt>
<dd><p>Array of IDs (e.g. c(299, 300, 302) for HDL, LDL, trigs)</p></dd>


<dt id="arg-clump-r-">clump_r2<a class="anchor" aria-label="anchor" href="#arg-clump-r-"></a></dt>
<dd><p>The default is <code>0.01</code>.</p></dd>


<dt id="arg-clump-kb">clump_kb<a class="anchor" aria-label="anchor" href="#arg-clump-kb"></a></dt>
<dd><p>The default is <code>10000</code>.</p></dd>


<dt id="arg-harmonise-strictness">harmonise_strictness<a class="anchor" aria-label="anchor" href="#arg-harmonise-strictness"></a></dt>
<dd><p>See the <code>action</code> option of <code><a href="harmonise_data.html">harmonise_data()</a></code>. The default is <code>2</code>.</p></dd>


<dt id="arg-opengwas-jwt">opengwas_jwt<a class="anchor" aria-label="anchor" href="#arg-opengwas-jwt"></a></dt>
<dd><p>Used to authenticate protected endpoints. Login to <a href="https://api.opengwas.io" class="external-link">https://api.opengwas.io</a> to obtain a jwt. Provide the jwt string here, or store in .Renviron under the keyname OPENGWAS_JWT.</p></dd>


<dt id="arg-find-proxies">find_proxies<a class="anchor" aria-label="anchor" href="#arg-find-proxies"></a></dt>
<dd><p>Look for proxies? This slows everything down but is more accurate. The default is <code>TRUE</code>.</p></dd>


<dt id="arg-force-server">force_server<a class="anchor" aria-label="anchor" href="#arg-force-server"></a></dt>
<dd><p>Whether to search through pre-clumped dataset or to re-extract and clump directly from the server. The default is <code>FALSE</code>.</p></dd>


<dt id="arg-pval-threshold">pval_threshold<a class="anchor" aria-label="anchor" href="#arg-pval-threshold"></a></dt>
<dd><p>Instrument detection p-value threshold. Default = <code>5e-8</code></p></dd>


<dt id="arg-pop">pop<a class="anchor" aria-label="anchor" href="#arg-pop"></a></dt>
<dd><p>Which 1000 genomes super population to use for clumping when using the server</p></dd>


<dt id="arg-plink-bin">plink_bin<a class="anchor" aria-label="anchor" href="#arg-plink-bin"></a></dt>
<dd><p>If <code>NULL</code> and <code>bfile</code> is not <code>NULL</code> then will detect packaged plink binary for specific OS. Otherwise specify path to plink binary. Default = <code>NULL</code></p></dd>


<dt id="arg-bfile">bfile<a class="anchor" aria-label="anchor" href="#arg-bfile"></a></dt>
<dd><p>If this is provided then will use the API. Default = <code>NULL</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data frame in <code>exposure_dat</code> format</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

