<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Reads in outcome data. Checks and organises columns for use with MR or enrichment tests.
Infers p-values when possible from beta and se."><title>Read outcome data — read_outcome_data • TwoSampleMR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Read outcome data — read_outcome_data"><meta property="og:description" content="Reads in outcome data. Checks and organises columns for use with MR or enrichment tests.
Infers p-values when possible from beta and se."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-guide">Guide</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-guide">
    <a class="dropdown-item" href="../articles/index.html">List of articles</a>
    <a class="dropdown-item" href="../articles/introduction.html">Introduction</a>
    <a class="dropdown-item" href="../articles/exposure.html">Exposure data</a>
    <a class="dropdown-item" href="../articles/outcome.html">Outcome data</a>
    <a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a>
    <a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a>
    <a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrcieu/TwoSampleMR">Source</a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Read outcome data</h1>
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/HEAD/R/read_data.R" class="external-link"><code>R/read_data.R</code></a></small>
      <div class="d-none name"><code>read_outcome_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads in outcome data. Checks and organises columns for use with MR or enrichment tests.
Infers p-values when possible from beta and se.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_outcome_data</span><span class="op">(</span></span>
<span>  <span class="va">filename</span>,</span>
<span>  snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">" "</span>,</span>
<span>  phenotype_col <span class="op">=</span> <span class="st">"Phenotype"</span>,</span>
<span>  snp_col <span class="op">=</span> <span class="st">"SNP"</span>,</span>
<span>  beta_col <span class="op">=</span> <span class="st">"beta"</span>,</span>
<span>  se_col <span class="op">=</span> <span class="st">"se"</span>,</span>
<span>  eaf_col <span class="op">=</span> <span class="st">"eaf"</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="st">"effect_allele"</span>,</span>
<span>  other_allele_col <span class="op">=</span> <span class="st">"other_allele"</span>,</span>
<span>  pval_col <span class="op">=</span> <span class="st">"pval"</span>,</span>
<span>  units_col <span class="op">=</span> <span class="st">"units"</span>,</span>
<span>  ncase_col <span class="op">=</span> <span class="st">"ncase"</span>,</span>
<span>  ncontrol_col <span class="op">=</span> <span class="st">"ncontrol"</span>,</span>
<span>  samplesize_col <span class="op">=</span> <span class="st">"samplesize"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  id_col <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>  min_pval <span class="op">=</span> <span class="fl">1e-200</span>,</span>
<span>  log_pval <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chr_col <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>  pos_col <span class="op">=</span> <span class="st">"pos"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>filename</dt>
<dd><p>Filename. Must have header with at least SNP column present.</p></dd>


<dt>snps</dt>
<dd><p>SNPs to extract. If <code>NULL</code>, which the default, then doesn't extract any and keeps all.</p></dd>


<dt>sep</dt>
<dd><p>Specify delimeter in file. The default is space, i.e. <code>sep=" "</code>.</p></dd>


<dt>phenotype_col</dt>
<dd><p>Optional column name for the column with phenotype name corresponding the the SNP. If not present then will be created with the value <code>"Outcome"</code>. Default is <code>"Phenotype"</code>.</p></dd>


<dt>snp_col</dt>
<dd><p>Required name of column with SNP rs IDs. The default is <code>"SNP"</code>.</p></dd>


<dt>beta_col</dt>
<dd><p>Required for MR. Name of column with effect sizes. THe default is <code>"beta"</code>.</p></dd>


<dt>se_col</dt>
<dd><p>Required for MR. Name of column with standard errors. The default is <code>"se"</code>.</p></dd>


<dt>eaf_col</dt>
<dd><p>Required for MR. Name of column with effect allele frequency. The default is <code>"eaf"</code>.</p></dd>


<dt>effect_allele_col</dt>
<dd><p>Required for MR. Name of column with effect allele. Must be "A", "C", "T" or "G". The default is <code>"effect_allele"</code>.</p></dd>


<dt>other_allele_col</dt>
<dd><p>Required for MR. Name of column with non effect allele. Must be "A", "C", "T" or "G". The default is <code>"other_allele"</code>.</p></dd>


<dt>pval_col</dt>
<dd><p>Required for enrichment tests. Name of column with p-value. The default is <code>"pval"</code>.</p></dd>


<dt>units_col</dt>
<dd><p>Optional column name for units. The default is <code>"units"</code>.</p></dd>


<dt>ncase_col</dt>
<dd><p>Optional column name for number of cases. The default is <code>"ncase"</code>.</p></dd>


<dt>ncontrol_col</dt>
<dd><p>Optional column name for number of controls. The default is <code>"ncontrol"</code>.</p></dd>


<dt>samplesize_col</dt>
<dd><p>Optional column name for sample size. The default is <code>"samplesize"</code>.</p></dd>


<dt>gene_col</dt>
<dd><p>Optional column name for gene name. The default is <code>"gene"</code>.</p></dd>


<dt>id_col</dt>
<dd><p>Optional column name to give the dataset an ID. Will be generated automatically if not provided for every trait / unit combination. The default is <code>"id"</code>.</p></dd>


<dt>min_pval</dt>
<dd><p>Minimum allowed p-value. The default is <code>1e-200</code>.</p></dd>


<dt>log_pval</dt>
<dd><p>The pval is -log10(P). The default is <code>FALSE</code>.</p></dd>


<dt>chr_col</dt>
<dd><p>Optional column name for chromosome. Default is <code>"chr"</code>.</p></dd>


<dt>pos_col</dt>
<dd><p>Optional column name for genetic position Default is <code>"pos"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>data frame</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

