<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Major changes to the IEU GWAS resources for 2020 • TwoSampleMR</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Major changes to the IEU GWAS resources for 2020">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide">
<li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Major changes to the IEU GWAS resources for 2020</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/vignettes/gwas2020.Rmd" class="external-link"><code>vignettes/gwas2020.Rmd</code></a></small>
      <div class="d-none name"><code>gwas2020.Rmd</code></div>
    </div>

    
    
<p>This document details changes and new features specifically relating
to the TwoSampleMR R package and the GWAS database behind it.</p>
<div class="section level2">
<h2 id="what-has-changed">What has changed<a class="anchor" aria-label="anchor" href="#what-has-changed"></a>
</h2>
<div class="section level3">
<h3 id="dataset-ids">Dataset IDs<a class="anchor" aria-label="anchor" href="#dataset-ids"></a>
</h3>
<p>We have made a new system for naming datasets, and all datasets are
organised into data batches. Either new datasets are uploaded one at a
time in which case they are added to the <code>ieu-a</code> data batch,
or there is a bulk upload in which case a new batch is created. For
example, <code>ukb-a</code> is a bulk upload of the first round of the
Neale lab UKBiobank GWAS, and <code>ukb-b</code> is the IEU GWAS
analysis of the UKBiobank data. In most cases, a dataset is then
numbered arbitrarily within the batch. For example, the Locke et al 2014
BMI analysis was previously known as <code>2</code>, and it is now known
as <code>ieu-a-2</code>.</p>
<p>There is backward compatibility built into the R packages that access
the data, so if you use an ‘old’ ID, it will automatically translate
that to the new one. But it will give you a warning, and we urge you to
update your scripts to reflect this change.</p>
</div>
<div class="section level3">
<h3 id="authentication">Authentication<a class="anchor" aria-label="anchor" href="#authentication"></a>
</h3>
<p>Previously you would automatically be asked to authenticate any query
to the database, through google. Now, we are making authentication
voluntary - something that you do at the start of a session only if you
need access to specific private datasets on the database. For the vast
majority of use cases this is not required.</p>
<p>Another change is that the R package that managed the authentication
has updated, and the file tokens generated are slightly different. For
full information on how to deal with this, see here: <a href="https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication" class="external-link uri">https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication</a></p>
</div>
<div class="section level3">
<h3 id="ukbiobank-data-has-been-curated">UKBiobank data has been curated<a class="anchor" aria-label="anchor" href="#ukbiobank-data-has-been-curated"></a>
</h3>
<p>We conducted a large GWAS analysis using a pipeline that
systematically analysed every <a href="https://pubmed.ncbi.nlm.nih.gov/29040602/" class="external-link">PHESANT</a> phenotype
in UK Biobank. There were previously ~20k traits with complete GWAS
data, but a majority of these were binary traits based on very few
numbers of cases. We have now filtered out unreliable datasets, there
are 2514 traits remaining, with any binary traits removed that had fewer
than 1000 cases. Another issue is the combination of small numbers of
cases and allele frequency - here minor allele count (MAC) for a
particular association could be very small which would lead to high
false positives when using Bolt-LMM. The remaining traits have been
filtered to only retain associations where the MAC &gt; 90.</p>
<p>Document detailing this investigation here: <a href="https://htmlpreview.github.io/?https://raw.githubusercontent.com/MRCIEU/ukbb-gwas-analysis/master/docs/ldsc_clumped_analysis.html?token=AAOV6TBQXEXEPT7SUXXLWMC6DWP3O" class="external-link uri">https://htmlpreview.github.io/?https://raw.githubusercontent.com/MRCIEU/ukbb-gwas-analysis/master/docs/ldsc_clumped_analysis.html?token=AAOV6TBQXEXEPT7SUXXLWMC6DWP3O</a></p>
</div>
<div class="section level3">
<h3 id="all-data-is-now-harmonised">All data is now harmonised<a class="anchor" aria-label="anchor" href="#all-data-is-now-harmonised"></a>
</h3>
<p>Previously the data were QC’d to remove malformed results and then
deposited as we found them. We are now also pre-harmonising all the
data. This means that all alleles are coded on the forward strand, and
the non-effect allele is always aligned to the human genome reference
sequence B37 (so the effect allele is the non-reference allele). This
does mean that sometimes variants have been removed if they did not map
to the human genome, and for most datasets the effect allele has been
switched for approximately half of all sites. When an effect allele
changes we do of course switch the sign of the effect size, so it should
not impact any MR results.</p>
</div>
<div class="section level3">
<h3 id="ld-reference-panel-is-now-harmonised">LD reference panel is now harmonised<a class="anchor" aria-label="anchor" href="#ld-reference-panel-is-now-harmonised"></a>
</h3>
<p>We have updated the LD reference panel to be harmonised against human
genome build 37, and as a consequence a few variants have been lost from
the version that was previously used.</p>
</div>
<div class="section level3">
<h3 id="instrument-lists-are-up-to-date">Instrument lists are up-to-date<a class="anchor" aria-label="anchor" href="#instrument-lists-are-up-to-date"></a>
</h3>
<p>Previously we were pre-clumping the tophits and storing them in the
MRInstruments R package, and there was often a delay in updating the
MRInstruments R package after new datasets were uploaded to the
database. We have moved away from this model. Everything dataset is
pre-clumped, but that is stored in the database. If you request default
clumping values when extracting the tophits of a dataset, it will still
be fast but it is retrieving the data from the server, and not from the
MRInstruments package. You can continue to use the MRInstruments package
for GWAS hits from e.g. GTEx or the EBI GWAS catalog.</p>
</div>
<div class="section level3">
<h3 id="dbsnp-rs-ids">dbSNP rs IDs<a class="anchor" aria-label="anchor" href="#dbsnp-rs-ids"></a>
</h3>
<p>All rs IDs have been mapped to dbSNP build 144. Therefore, some rs
IDs may have changed, but there is stronger alignment across all
datasets.</p>
</div>
<div class="section level3">
<h3 id="everything-is-faster">Everything is faster<a class="anchor" aria-label="anchor" href="#everything-is-faster"></a>
</h3>
<p>We are using Elasticsearch and Neo4j on an Oracle Cloud
Infrastructure to serve the data. It’s much faster. Interestingly, it
actually gets faster when more people are using it because the cache
gets ‘warmed up’ by more requests.</p>
</div>
</div>
<div class="section level2">
<h2 id="what-is-new">What is new<a class="anchor" aria-label="anchor" href="#what-is-new"></a>
</h2>
<div class="section level3">
<h3 id="browse-available-datasets-online">Browse available datasets online<a class="anchor" aria-label="anchor" href="#browse-available-datasets-online"></a>
</h3>
<p>We have a new home for the GWAS summary data: <a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a>.</p>
</div>
<div class="section level3">
<h3 id="chromosome-and-position">Chromosome and position<a class="anchor" aria-label="anchor" href="#chromosome-and-position"></a>
</h3>
<p>All variants have been mapped to chromosome and position
(hg19/build37). You can query based on chromosome position coordinates.
This means either a list of <code>&lt;chr:pos&gt;</code> values, or a
list of <code>&lt;chr:pos1-pos2&gt;</code> ranges.</p>
</div>
<div class="section level3">
<h3 id="indels-are-retained">INDELs are retained<a class="anchor" aria-label="anchor" href="#indels-are-retained"></a>
</h3>
<p>Previously we were excluding these, but they are now retained</p>
</div>
<div class="section level3">
<h3 id="multi-allelic-variants-are-retained">Multi-allelic variants are retained<a class="anchor" aria-label="anchor" href="#multi-allelic-variants-are-retained"></a>
</h3>
<p>Previously we were excluding these, but they are now retained. Be
warned that if you extract a variant that has multiple alleles then you
may get more than one row for that variant.</p>
</div>
<div class="section level3">
<h3 id="more-data">More data<a class="anchor" aria-label="anchor" href="#more-data"></a>
</h3>
<p>Automated download from the EBI repository, and an automated upload
system and batch data processing system means that more data can be
added faster to keep the database current.</p>
</div>
<div class="section level3">
<h3 id="error-messages-are-more-informative">Error messages are more informative<a class="anchor" aria-label="anchor" href="#error-messages-are-more-informative"></a>
</h3>
<p>Previously if a query to the database failed it didn’t give a reason,
hopefully there is more clarity regarding what is happening now. You can
also check the status of the server here: <a href="https://api.opengwas.io/api/" class="external-link uri">https://api.opengwas.io/api/</a></p>
</div>
<div class="section level3">
<h3 id="easier-programmatic-access-to-the-database">Easier programmatic access to the database<a class="anchor" aria-label="anchor" href="#easier-programmatic-access-to-the-database"></a>
</h3>
<p>We are trying to make it as flexible as possible to access the data.
The TwoSampleMR R package was previously the only programmatic way to
access the data, now we have the following options:</p>
<ul>
<li>
<a href="https://mrcieu.github.io/ieugwasr/" class="external-link">ieugwasr R package</a>:
All the TwoSampleMR functions that access the data are done by calling
this package now. It is a simple wrapper around the API that controls
access to the database.</li>
<li>
<a href="https://github.com/MRCIEU/ieugwaspy/" class="external-link">ieugwaspy python
package</a>: Similar functionality to ieugwasr (Under
construction).</li>
<li>
<a href="https://api.opengwas.io/api/" class="external-link">API</a>: You can use the API
directly, e.g. for building your own services or applications.</li>
</ul>
</div>
<div class="section level3">
<h3 id="local-ld-operations">Local LD operations<a class="anchor" aria-label="anchor" href="#local-ld-operations"></a>
</h3>
<p>It is now possible to perform clumping, or create LD matrices, using
your own local LD reference dataset. You can download the one that we
have been using here: <a href="https://github.com/mrcieu/gwasglue#reference-datasets" class="external-link uri">https://github.com/mrcieu/gwasglue#reference-datasets</a>,
or create your own plink format dataset e.g. with larger samples or for
different ancestries. See the LD clumping functions in the <a href="https://mrcieu.github.io/ieugwasr/" class="external-link">ieugwasr</a> package for more
details.</p>
</div>
<div class="section level3">
<h3 id="access-the-data-directly">Access the data directly<a class="anchor" aria-label="anchor" href="#access-the-data-directly"></a>
</h3>
<p>Previously the data was only accessible through the database. Now the
data can be downloaded in “GWAS VCF” format from here <a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a>. (IEU members can access all
the data on RDSF or bluecrystal4 directly). This means that if you want
to perform very large or numerous operations, you can do it on HPC or
locally in a more performant manner by using the data files directly.
Please see the <a href="https://github.com/mrcieu/gwasvcf" class="external-link">gwasvcf R
package</a> on how to work with these data.</p>
</div>
<div class="section level3">
<h3 id="connect-the-data-to-different-analytical-tools">Connect the data to different analytical tools<a class="anchor" aria-label="anchor" href="#connect-the-data-to-different-analytical-tools"></a>
</h3>
<p>Either the data in the database, or the GWAS VCF files, can be
queried and the results translated into the formats for a bunch of
different R packages for MR, colocalisation, fine mapping, etc. Have a
look at the <a href="https://github.com/mrcieu/gwasglue" class="external-link">gwasglue R
package</a>, to see what is available and how to do this. It’s still
under construction, but feel free to try it, make suggestions, and
contribute code.</p>
</div>
</div>
<div class="section level2">
<h2 id="key-links">Key links<a class="anchor" aria-label="anchor" href="#key-links"></a>
</h2>
<ul>
<li>The IEU GWAS database: <a href="https://gwas.mrcieu.ac.uk" class="external-link uri">https://gwas.mrcieu.ac.uk</a>
</li>
<li>API to the IEU GWAS database: <a href="https://api.opengwas.io/api/" class="external-link uri">https://api.opengwas.io/api/</a>
</li>
<li>ieugwasr package, for R access to the API: <a href="https://mrcieu.github.io/ieugwasr/" class="external-link uri">https://mrcieu.github.io/ieugwasr/</a>
</li>
<li>ieugwaspy package, for python access to the API: <a href="https://github.com/MRCIEU/ieugwaspy/" class="external-link uri">https://github.com/MRCIEU/ieugwaspy/</a> (Under
construction)</li>
<li>gwasvcf package, R interface to GWAS VCF files: <a href="https://mrcieu.github.io/gwasvcf/" class="external-link uri">https://mrcieu.github.io/gwasvcf/</a>
</li>
<li>pygwasvcf package, python interface to GWAS VCF files: <a href="https://github.com/mrcieu/pygwasvcf" class="external-link uri">https://github.com/mrcieu/pygwasvcf</a> (Under
construction)</li>
<li>gwasglue package, linking GWAS data to various analytical methods:
<a href="https://mrcieu.github.io/gwasglue/" class="external-link uri">https://mrcieu.github.io/gwasglue/</a> (Functional, under
construction)</li>
<li>gwas2vcf online tool, allowing users to create GWAS VCF files from
summary data: <a href="https://github.com/MRCIEU/gwas2vcf" class="external-link uri">https://github.com/MRCIEU/gwas2vcf</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="how-to-request-new-data">How to request new data<a class="anchor" aria-label="anchor" href="#how-to-request-new-data"></a>
</h2>
<p>We have setup a github issues page here: <a href="https://github.com/MRCIEU/opengwas-requests/issues" class="external-link uri">https://github.com/MRCIEU/opengwas-requests/issues</a></p>
<p>Please visit here to make a log of new data requests, or to
contribute new data.</p>
</div>
<div class="section level2">
<h2 id="backwards-compatibility">Backwards compatibility<a class="anchor" aria-label="anchor" href="#backwards-compatibility"></a>
</h2>
<p>To install the new version of TwoSampleMR, perform as normal:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"MRCIEU/TwoSampleMR"</span><span class="op">)</span></span></code></pre></div>
<p>To update the package just run the
<code>remotes::install_github("MRCIEU/TwoSampleMR")</code> command
again.</p>
<p>We recommend using this new version going forwards but for a limited
time we are enabling backwards compatibility, in case you are in the
middle of analysis or need to reproduce old analysis. In order to use
the legacy version of the package and the database, install using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"MRCIEU/TwoSampleMR@0.4.26"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
