<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Harmonise data • TwoSampleMR</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Harmonise data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.20</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide">
<li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Harmonise data</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/vignettes/harmonise.Rmd" class="external-link"><code>vignettes/harmonise.Rmd</code></a></small>
      <div class="d-none name"><code>harmonise.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/TwoSampleMR" class="external-link">TwoSampleMR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The exposure data and outcome data are now obtained, e.g.:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bmi_exp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_instruments.html">extract_instruments</a></span><span class="op">(</span>outcomes <span class="op">=</span> <span class="st">'ieu-a-2'</span><span class="op">)</span></span>
<span><span class="va">chd_out_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span>snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>, outcomes <span class="op">=</span> <span class="st">'ieu-a-7'</span><span class="op">)</span></span></code></pre></div>
<p>but it is important to harmonise the effects. This means that the
effect of a SNP on the exposure and the effect of that SNP on the
outcome must each correspond to the same allele.</p>
<p><strong>Note:</strong> The IEU GWAS database contains data that is
already harmonised, meaning that the non-effect allele is aligned to the
human genome reference sequence (build 37). It’s still recommended to
harmonise, but in principle everything should be on the forward strand
and effect alleles always relating to the same allele. Some
discrepancies could arise if there are multi-allelic variants that are
represented as different bi-allelic variants in different studies.</p>
<p>To harmonise the exposure and outcome data, do the following:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/harmonise_data.html">harmonise_data</a></span><span class="op">(</span></span>
<span>    exposure_dat <span class="op">=</span> <span class="va">bmi_exp_dat</span>,</span>
<span>    outcome_dat <span class="op">=</span> <span class="va">chd_out_dat</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Harmonising Body mass index || id:ieu-a-2 (ieu-a-2) and Coronary heart disease || id:ieu-a-7 (ieu-a-7)</span></span></code></pre></div>
<p>This creates a new data frame that has the exposure data and outcome
data combined.</p>
<p>If there were 3 exposure traits and 3 outcome traits then there will
be 9 sets of harmonisations being performed - harmonising the SNP
effects of exposure trait 1 against outcome trait 1; exposure trait 1
against outcome trait 2; and so on.</p>
</div>
<div class="section level2">
<h2 id="dealing-with-strand-issues">Dealing with strand issues<a class="anchor" aria-label="anchor" href="#dealing-with-strand-issues"></a>
</h2>
<p>Recent GWASs typically present the effects of a SNP in reference to
the allele on the forward strand. But as reference panels are updated
the forward strand sometimes changes, and GWASs from a few years ago
aren’t guaranteed to be using forward strand conventions.</p>
<p>Some examples are shown below:</p>
<div class="section level3">
<h3 id="correct-unambiguous">Correct, unambiguous<a class="anchor" aria-label="anchor" href="#correct-unambiguous"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = 0.05
effect allele = A
other allele = G</code></pre>
<p>Here the effect allele on the exposure and the outcome is the
same</p>
</div>
<div class="section level3">
<h3 id="incorrect-reference-unambiguous">Incorrect reference, unambiguous<a class="anchor" aria-label="anchor" href="#incorrect-reference-unambiguous"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = -0.05
effect allele = C
other allele = T</code></pre>
<p>Here the outcome GWAS is presenting the effect for the alternate
allele on the reverse strand. We need to flip the outcome effect to 0.05
to correspond to the same allele as the exposure GWAS on the forward
strand.</p>
</div>
<div class="section level3">
<h3 id="ambiguous">Ambiguous<a class="anchor" aria-label="anchor" href="#ambiguous"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = -0.05
effect allele = A
other allele = C</code></pre>
<p>Here the alleles do not correspond for the same SNP, so this SNP will
be discarded from the analysis.</p>
</div>
<div class="section level3">
<h3 id="palindromic-snp-inferrable">Palindromic SNP, inferrable<a class="anchor" aria-label="anchor" href="#palindromic-snp-inferrable"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = T
effect allele frequency = 0.11

outcome effect = -0.05
effect allele = A
other allele = T
effect allele frequency = 0.91</code></pre>
<p>Here the alleles correspond, but it is a palindromic SNP, such that
the alleles on the forward strand are the same as on the reverse strand
(A/T on forward is T/A on the reverse). However, the allele frequency of
the effect allele gives us information - if the outcome effect allele
(A) were on the forward strand we would expect it to have a low allele
frequency, but given it has a high frequency (0.91) we infer that the
outcome GWAS is presenting the effect on the reverse strand for the
alternative allele. We would flip the effect to 0.05 for the outcome
GWAS.</p>
</div>
<div class="section level3">
<h3 id="palindromic-snp-not-inferrable">Palindromic SNP, not inferrable<a class="anchor" aria-label="anchor" href="#palindromic-snp-not-inferrable"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = T
effect allele frequency = 0.50

outcome effect = -0.05
effect allele = A
other allele = T
effect allele frequency = 0.50</code></pre>
<p>This is similar to the above, except the allele frequency no longer
gives us information about the strand. We would discard this SNP. This
is done for any palindromic SNPs that have minor allele frequency above
0.42.</p>
</div>
<div class="section level3">
<h3 id="options">Options<a class="anchor" aria-label="anchor" href="#options"></a>
</h3>
<p>There are three options to harmonising the data.</p>
<ol style="list-style-type: decimal">
<li>Assume all alleles are presented on the forward strand</li>
<li>Try to infer the forward strand alleles using allele frequency
information</li>
<li>Correct the strand for non-palindromic SNPs, but drop all
palindromic SNPs</li>
</ol>
<p>By default, the <code>harmonise_data</code> function uses option 2,
but this can be modified using the <code>action</code> argument,
e.g. <code>harmonise_data(exposure_dat, outcome_dat, action = 3)</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="drop-duplicate-exposure-outcome-summary-sets">Drop duplicate exposure-outcome summary sets<a class="anchor" aria-label="anchor" href="#drop-duplicate-exposure-outcome-summary-sets"></a>
</h2>
<p>After data harmonisation, users may find that their dataset contains
duplicate exposure-outcome summary sets. This can arise, for example,
when a GWAS consortium has released multiple results from separate GWAS
analyses for the same trait. For example, there are multiple GWAS
summary datasets for body mass index and coronary heart disease:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/available_outcomes.html">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span><span class="op">[</span><span class="va">ao</span><span class="op">$</span><span class="va">trait</span> <span class="op">==</span> <span class="st">"Body mass index"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trait"</span>, <span class="st">"id"</span>, <span class="st">"pmid"</span>, <span class="st">"author"</span>, <span class="st">"sample_size"</span>, <span class="st">"nsnp"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;                 trait                 id     pmid                    author</span></span>
<span><span class="co">#&gt; 3958  Body mass index ebi-a-GCST90103751 35051171                   Wong HS</span></span>
<span><span class="co">#&gt; 4015  Body mass index ebi-a-GCST90095039 35399580 Fern&lt;U+00E1&gt;ndez-Rhodes L</span></span>
<span><span class="co">#&gt; 4020  Body mass index ebi-a-GCST90095034 35399580 Fern&lt;U+00E1&gt;ndez-Rhodes L</span></span>
<span><span class="co">#&gt; 6032  Body mass index ebi-a-GCST90029007 29892013                    Loh PR</span></span>
<span><span class="co">#&gt; 6821  Body mass index ebi-a-GCST90025994 34226706                 Barton AR</span></span>
<span><span class="co">#&gt; 7045  Body mass index ebi-a-GCST90018947 34594039                  Sakaue S</span></span>
<span><span class="co">#&gt; 7259  Body mass index ebi-a-GCST90018727 34594039                  Sakaue S</span></span>
<span><span class="co">#&gt; 10738 Body mass index           ieu-a-94 23754948                Randall JC</span></span>
<span><span class="co">#&gt; 12717 Body mass index            ieu-a-2 25673413                  Locke AE</span></span>
<span><span class="co">#&gt; 14254 Body mass index           ieu-a-95 23754948                Randall JC</span></span>
<span><span class="co">#&gt; 16676 Body mass index          ieu-a-974 25673413                  Locke AE</span></span>
<span><span class="co">#&gt; 19343 Body mass index            bbj-a-3 28892062                Ishigaki K</span></span>
<span><span class="co">#&gt; 26475 Body mass index   ebi-a-GCST006368 30108127               Hoffmann TJ</span></span>
<span><span class="co">#&gt; 28065 Body mass index         ieu-b-4815       NA                   Howe LJ</span></span>
<span><span class="co">#&gt; 28419 Body mass index            bbj-a-2 28892062                Ishigaki K</span></span>
<span><span class="co">#&gt; 32371 Body mass index         ieu-b-4816       NA                   Howe LJ</span></span>
<span><span class="co">#&gt; 33869 Body mass index          ieu-a-785 25673413                  Locke AE</span></span>
<span><span class="co">#&gt; 39217 Body mass index   ebi-a-GCST002783 25673413                  Locke AE</span></span>
<span><span class="co">#&gt; 40065 Body mass index            bbj-a-1 28892062                Ishigaki K</span></span>
<span><span class="co">#&gt; 43262 Body mass index   ebi-a-GCST004904 28892062                 Akiyama M</span></span>
<span><span class="co">#&gt; 43743 Body mass index   ebi-a-GCST006802 26961502                   Wood AR</span></span>
<span><span class="co">#&gt; 47894 Body mass index          ieu-a-835 25673413                  Locke AE</span></span>
<span><span class="co">#&gt; 48917 Body mass index   ebi-a-GCST008025 31217584                 Wojcik GL</span></span>
<span><span class="co">#&gt; 49208 Body mass index         ieu-a-1089 26961502                      Wood</span></span>
<span><span class="co">#&gt;       sample_size     nsnp</span></span>
<span><span class="co">#&gt; 3958        21930  6370138</span></span>
<span><span class="co">#&gt; 4015       330793  2401077</span></span>
<span><span class="co">#&gt; 4020        56161  8764141</span></span>
<span><span class="co">#&gt; 6032       532396 11973091</span></span>
<span><span class="co">#&gt; 6821       457756  4238669</span></span>
<span><span class="co">#&gt; 7045       359983 19066885</span></span>
<span><span class="co">#&gt; 7259       163835 12502877</span></span>
<span><span class="co">#&gt; 10738       60586  2736876</span></span>
<span><span class="co">#&gt; 12717      339224  2555511</span></span>
<span><span class="co">#&gt; 14254       73137  2736876</span></span>
<span><span class="co">#&gt; 16676      171977  2494613</span></span>
<span><span class="co">#&gt; 19343       72390  6108953</span></span>
<span><span class="co">#&gt; 26475      315347 27854527</span></span>
<span><span class="co">#&gt; 28065       51852       NA</span></span>
<span><span class="co">#&gt; 28419       85894  6108953</span></span>
<span><span class="co">#&gt; 32371       99998  7191606</span></span>
<span><span class="co">#&gt; 33869      152893  2477659</span></span>
<span><span class="co">#&gt; 39217      236781  2529499</span></span>
<span><span class="co">#&gt; 40065      158284  5961600</span></span>
<span><span class="co">#&gt; 43262      158284  5952516</span></span>
<span><span class="co">#&gt; 43743      119688  8580466</span></span>
<span><span class="co">#&gt; 47894      322154  2554668</span></span>
<span><span class="co">#&gt; 48917       21955 34343880</span></span>
<span><span class="co">#&gt; 49208      120286  8654252</span></span>
<span><span class="va">ao</span><span class="op">[</span><span class="va">ao</span><span class="op">$</span><span class="va">trait</span> <span class="op">==</span> <span class="st">"Coronary heart disease"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trait"</span>, <span class="st">"id"</span>, <span class="st">"pmid"</span>, <span class="st">"author"</span>, <span class="st">"ncase"</span>, <span class="st">"ncontrol"</span>, <span class="st">"nsnp"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;                        trait               id     pmid      author ncase</span></span>
<span><span class="co">#&gt; 14897 Coronary heart disease          ieu-a-7 26343387      Nikpay 60801</span></span>
<span><span class="co">#&gt; 23614 Coronary heart disease          ieu-a-9 23202125    Deloukas 63746</span></span>
<span><span class="co">#&gt; 27414 Coronary heart disease ebi-a-GCST000998 21378990 Schunkert H 22233</span></span>
<span><span class="co">#&gt; 38602 Coronary heart disease          ieu-a-8 21378990 Schunkert H 22233</span></span>
<span><span class="co">#&gt; 45294 Coronary heart disease          ieu-a-6 21378988       Peden 15420</span></span>
<span><span class="co">#&gt;       ncontrol    nsnp</span></span>
<span><span class="co">#&gt; 14897   123504 9455779</span></span>
<span><span class="co">#&gt; 23614   130681   79129</span></span>
<span><span class="co">#&gt; 27414    64762 2415020</span></span>
<span><span class="co">#&gt; 38602    64762 2420361</span></span>
<span><span class="co">#&gt; 45294    15062  540233</span></span></code></pre></div>
<p>There are therefore multiple potential combinations of body mass
index and coronary heart disease, which would likely lead to duplicate
MR analyses. We recommend that users prune their datasets so that only
the exposure-outcome combination with the highest expected power is
retained. This can be done by selecting the exposure-outcome summary set
with the largest sample size for the outcome, using the power_prune
function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/power_prune.html">power_prune</a></span><span class="op">(</span><span class="va">dat</span>, method <span class="op">=</span> <span class="fl">1</span>, dist.outcome <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span></span></code></pre></div>
<p>This drops the duplicate exposure-outcome sets with the smaller
outcome sample size (number of cases for binary outcomes). Remaining
duplicates are then dropped on the basis of the exposure sample size.
However, if there are a large number of SNPs available to instrument an
exposure, the outcome GWAS with the better SNP coverage may provide
better power than the outcome GWAS with the larger sample size. This can
occur, for example, if the larger outcome GWAS has used a targeted
genotyping array. In such instances, it may be better to prune studies
on the basis of instrument strength (i.e. variation in exposure
explained by the instrumental SNPs) as well as sample size. This can be
done by setting the method argument to 2:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/power_prune.html">power_prune</a></span><span class="op">(</span><span class="va">dat</span>, method <span class="op">=</span> <span class="fl">2</span>, dist.outcome <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span></span></code></pre></div>
<p>This procedure drops duplicate exposure-outcome sets on the basis of
instrument strength and sample size, and assumes that the SNP-exposure
effects correspond to a continuous trait with a normal distribution
(i.e. exposure should not be binary). The SNP-outcome effects can
correspond to either a binary or continuous trait (default behaviour is
to assume a binary distribution). If the exposure is binary then method
1 should be used.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
