<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="TwoSampleMR">
<title>Harmonise data • TwoSampleMR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Harmonise data">
<meta property="og:description" content="TwoSampleMR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-guide">Guide</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-guide">
    <a class="dropdown-item" href="../articles/index.html">List of articles</a>
    <a class="dropdown-item" href="../articles/introduction.html">Introduction</a>
    <a class="dropdown-item" href="../articles/exposure.html">Exposure data</a>
    <a class="dropdown-item" href="../articles/outcome.html">Outcome data</a>
    <a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a>
    <a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a>
    <a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrcieu/TwoSampleMR">Source</a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Harmonise data</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/HEAD/vignettes/harmonise.Rmd" class="external-link"><code>vignettes/harmonise.Rmd</code></a></small>
      <div class="d-none name"><code>harmonise.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/TwoSampleMR" class="external-link">TwoSampleMR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The exposure data and outcome data are now obtained, e.g.:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/available_outcomes.html">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; API: public: http://gwas-api.mrcieu.ac.uk/</span></span>
<span><span class="va">bmi_exp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_instruments.html">extract_instruments</a></span><span class="op">(</span>outcomes <span class="op">=</span> <span class="st">'ieu-a-2'</span><span class="op">)</span></span>
<span><span class="va">chd_out_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span>snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>, outcomes <span class="op">=</span> <span class="st">'ieu-a-7'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Extracting data for 79 SNP(s) from 1 GWAS(s)</span></span></code></pre></div>
<p>but it is important to harmonise the effects. This means that the
effect of a SNP on the exposure and the effect of that SNP on the
outcome must each correspond to the same allele.</p>
<p><strong>Note:</strong> The IEU GWAS database contains data that is
already harmonised, meaning that the non-effect allele is aligned to the
human genome reference sequence (build 37). It’s still recommended to
harmonise, but in principle everything should be on the forward strand
and effect alleles always relating to the same allele. Some
discrepancies could arise if there are multi-allelic variants that are
represented as different bi-allelic variants in different studies.</p>
<p>To harmonise the exposure and outcome data, do the following:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/harmonise_data.html">harmonise_data</a></span><span class="op">(</span></span>
<span>    exposure_dat <span class="op">=</span> <span class="va">bmi_exp_dat</span>, </span>
<span>    outcome_dat <span class="op">=</span> <span class="va">chd_out_dat</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Harmonising Body mass index || id:ieu-a-2 (ieu-a-2) and Coronary heart disease || id:ieu-a-7 (ieu-a-7)</span></span></code></pre></div>
<p>This creates a new data frame that has the exposure data and outcome
data combined.</p>
<p>If there were 3 exposure traits and 3 outcome traits then there will
be 9 sets of harmonisations being performed - harmonising the SNP
effects of exposure trait 1 against outcome trait 1; exposure trait 1
against outcome trait 2; and so on.</p>
</div>
<div class="section level2">
<h2 id="dealing-with-strand-issues">Dealing with strand issues<a class="anchor" aria-label="anchor" href="#dealing-with-strand-issues"></a>
</h2>
<p>Recent GWASs typically present the effects of a SNP in reference to
the allele on the forward strand. But as reference panels are updated
the forward strand sometimes changes, and GWASs from a few years ago
aren’t guaranteed to be using forward strand conventions.</p>
<p>Some examples are shown below:</p>
<div class="section level3">
<h3 id="correct-unambigious">Correct, unambigious<a class="anchor" aria-label="anchor" href="#correct-unambigious"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = 0.05
effect allele = A
other allele = G</code></pre>
<p>Here the effect allele on the exposure and the outcome is the
same</p>
</div>
<div class="section level3">
<h3 id="incorrect-reference-unambigious">Incorrect reference, unambigious<a class="anchor" aria-label="anchor" href="#incorrect-reference-unambigious"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = -0.05
effect allele = C
other allele = T</code></pre>
<p>Here the outcome GWAS is presenting the effect for the alternate
allele on the reverse strand. We need to flip the outcome effect to 0.05
to correspond to the same allele as the exposure GWAS on the forward
strand.</p>
</div>
<div class="section level3">
<h3 id="ambiguous">Ambiguous<a class="anchor" aria-label="anchor" href="#ambiguous"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = G

outcome effect = -0.05
effect allele = A
other allele = C</code></pre>
<p>Here the alleles do not correspond for the same SNP, so this SNP will
be discarded from the analysis.</p>
</div>
<div class="section level3">
<h3 id="palindromic-snp-inferrable">Palindromic SNP, inferrable<a class="anchor" aria-label="anchor" href="#palindromic-snp-inferrable"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = T
effect allele frequency = 0.11

outcome effect = -0.05
effect allele = A
other allele = T
effect allele frequency = 0.91</code></pre>
<p>Here the alleles correspond, but it is a palindromic SNP, such that
the alleles on the forward strand are the same as on the reverse strand
(A/T on forward is T/A on the reverse). However, the allele frequency of
the effect allele gives us information - if the outcome effect allele
(A) were on the forward strand we would expect it to have a low allele
frequency, but given it has a high frequency (0.91) we infer that the
outcome GWAS is presenting the effect on the reverse strand for the
alternative allele. We would flip the effect to 0.05 for the outcome
GWAS.</p>
</div>
<div class="section level3">
<h3 id="palindromic-snp-not-inferrable">Palindromic SNP, not inferrable<a class="anchor" aria-label="anchor" href="#palindromic-snp-not-inferrable"></a>
</h3>
<pre><code>exposure effect = 0.5
effect allele = A
other allele = T
effect allele frequency = 0.50

outcome effect = -0.05
effect allele = A
other allele = T
effect allele frequency = 0.50</code></pre>
<p>This is similar to the above, except the allele frequency no longer
gives us information about the strand. We would discard this SNP. This
is done for any palindromic SNPs that have minor allele frequency above
0.42.</p>
</div>
<div class="section level3">
<h3 id="options">Options<a class="anchor" aria-label="anchor" href="#options"></a>
</h3>
<p>There are three options to harmonising the data.</p>
<ol style="list-style-type: decimal">
<li>Assume all alleles are presented on the forward strand</li>
<li>Try to infer the forward strand alleles using allele frequency
information</li>
<li>Correct the strand for non-palindromic SNPs, but drop all
palindromic SNPs</li>
</ol>
<p>By default, the <code>harmonise_data</code> function uses option 2,
but this can be modified using the <code>action</code> argument,
e.g. <code>harmonise_data(exposure_dat, outcome_dat, action = 3)</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="drop-duplicate-exposure-outcome-summary-sets">Drop duplicate exposure-outcome summary sets<a class="anchor" aria-label="anchor" href="#drop-duplicate-exposure-outcome-summary-sets"></a>
</h2>
<p>After data harmonisation, users may find that their dataset contains
duplicate exposure-outcome summary sets. This can arise, for example,
when a GWAS consortium has released multiple results from separate GWAS
analyses for the same trait. For example, there are multiple GWAS
summary datasets for body mass index and coronary heart disease:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span><span class="op">[</span><span class="va">ao</span><span class="op">$</span><span class="va">trait</span> <span class="op">==</span> <span class="st">"Body mass index"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trait"</span>, <span class="st">"id"</span>, <span class="st">"pmid"</span>, <span class="st">"author"</span>, <span class="st">"sample_size"</span>, <span class="st">"nsnp"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 24 × 6</span></span></span>
<span><span class="co">#&gt;    trait           id                     pmid author         sample_size   nsnp</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Body mass index ebi-a-GCST90103751 35<span style="text-decoration: underline;">051</span>171 Wong HS              <span style="text-decoration: underline;">21</span>930 6.37<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Body mass index ebi-a-GCST90095039 35<span style="text-decoration: underline;">399</span>580 Fern&lt;U+00E1&gt;n…      <span style="text-decoration: underline;">330</span>793 2.40<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Body mass index ebi-a-GCST90095034 35<span style="text-decoration: underline;">399</span>580 Fern&lt;U+00E1&gt;n…       <span style="text-decoration: underline;">56</span>161 8.76<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Body mass index ebi-a-GCST90029007 29<span style="text-decoration: underline;">892</span>013 Loh PR              <span style="text-decoration: underline;">532</span>396 1.20<span style="color: #949494;">e</span>7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Body mass index ebi-a-GCST90025994 34<span style="text-decoration: underline;">226</span>706 Barton AR           <span style="text-decoration: underline;">457</span>756 4.24<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Body mass index ebi-a-GCST90018947 34<span style="text-decoration: underline;">594</span>039 Sakaue S            <span style="text-decoration: underline;">359</span>983 1.91<span style="color: #949494;">e</span>7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Body mass index ebi-a-GCST90018727 34<span style="text-decoration: underline;">594</span>039 Sakaue S            <span style="text-decoration: underline;">163</span>835 1.25<span style="color: #949494;">e</span>7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Body mass index ieu-a-94           23<span style="text-decoration: underline;">754</span>948 Randall JC           <span style="text-decoration: underline;">60</span>586 2.74<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Body mass index ieu-a-2            25<span style="text-decoration: underline;">673</span>413 Locke AE            <span style="text-decoration: underline;">339</span>224 2.56<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Body mass index ieu-a-95           23<span style="text-decoration: underline;">754</span>948 Randall JC           <span style="text-decoration: underline;">73</span>137 2.74<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 14 more rows</span></span></span>
<span><span class="va">ao</span><span class="op">[</span><span class="va">ao</span><span class="op">$</span><span class="va">trait</span> <span class="op">==</span> <span class="st">"Coronary heart disease"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trait"</span>, <span class="st">"id"</span>, <span class="st">"pmid"</span>, <span class="st">"author"</span>, <span class="st">"ncase"</span>, <span class="st">"ncontrol"</span>, <span class="st">"nsnp"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 7</span></span></span>
<span><span class="co">#&gt;   trait                  id                   pmid author  ncase ncontrol   nsnp</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Coronary heart disease ieu-a-7          26<span style="text-decoration: underline;">343</span>387 Nikpay  <span style="text-decoration: underline;">60</span>801   <span style="text-decoration: underline;">123</span>504 9.46<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Coronary heart disease ieu-a-9          23<span style="text-decoration: underline;">202</span>125 Delouk… <span style="text-decoration: underline;">63</span>746   <span style="text-decoration: underline;">130</span>681 7.91<span style="color: #949494;">e</span>4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Coronary heart disease ebi-a-GCST000998 21<span style="text-decoration: underline;">378</span>990 Schunk… <span style="text-decoration: underline;">22</span>233    <span style="text-decoration: underline;">64</span>762 2.42<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Coronary heart disease ieu-a-8          21<span style="text-decoration: underline;">378</span>990 Schunk… <span style="text-decoration: underline;">22</span>233    <span style="text-decoration: underline;">64</span>762 2.42<span style="color: #949494;">e</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Coronary heart disease ieu-a-6          21<span style="text-decoration: underline;">378</span>988 Peden   <span style="text-decoration: underline;">15</span>420    <span style="text-decoration: underline;">15</span>062 5.40<span style="color: #949494;">e</span>5</span></span></code></pre></div>
<p>There are therefore multiple potential combinations of body mass
index and coronary heart disease, which would likely lead to duplicate
MR analyses. We recommend that users prune their datasets so that only
the exposure-outcome combination with the highested expected power is
retained. This can be done by selecting the exposure-outcome summary set
with the largest sample size for the outcome, using the power_prune
function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/power_prune.html">power_prune</a></span><span class="op">(</span><span class="va">dat</span>, method <span class="op">=</span> <span class="fl">1</span>, dist.outcome <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span></span></code></pre></div>
<p>This drops the duplicate exposure-outcome sets with the smaller
outcome sample size (number of cases for binary outcomes). Remaining
duplicates are then dropped on the basis of the exposure sample size.
However, if there are a large number of SNPs available to instrument an
exposure, the outcome GWAS with the better SNP coverage may provide
better power than the outcome GWAS with the larger sample size. This can
occur, for example, if the larger outcome GWAS has used a targeted
genotyping array. In such instances, it may be better to prune studies
on the basis of instrument strength (i.e. variation in exposure
explained by the instrumental SNPs) as well as sample size. This can be
done by setting the method argument to 2:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/power_prune.html">power_prune</a></span><span class="op">(</span><span class="va">dat</span>, method <span class="op">=</span> <span class="fl">2</span>, dist.outcome <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span></span></code></pre></div>
<p>This procedure drops duplicate exposure-outcome sets on the basis of
instrument strength and sample size, and assumes that the SNP-exposure
effects correspond to a continuous trait with a normal distribution
(i.e. exposure should not be binary). The SNP-outcome effects can
correspond to either a binary or continuous trait (default behaviour is
to assume a binary distribution). If the exposure is binary then method
1 should be used.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
