<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction • TwoSampleMR</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.13</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide">
<li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>Two sample Mendelian randomisation (2SMR) is a method to estimate the
causal effect of an exposure on an outcome using only summary statistics
from genome wide association studies (GWAS). Though conceptually
straightforward, there are a number of steps that are required to
perform the analysis properly, and they can be cumbersome. The
TwoSampleMR package aims to make this easy by combining three important
components</p>
<ul>
<li>data management and harmonisation</li>
<li>the statistical routines to estimate the causal effects</li>
<li>connection to a large repository of the actual GWAS summary
statistics needed to perform the analyses.</li>
</ul>
<p>The general principles <span class="citation">(G. Davey Smith and
Ebrahim 2003; George Davey Smith and Hemani 2014)</span>, and
statistical methods <span class="citation">(Pierce and Burgess 2013;
Bowden, Davey Smith, and Burgess 2015)</span> can be found elsewhere,
here we will just outline how to use the R package.</p>
<p>This package uses the <a href="https://github.com/mrcieu/ieugwasr" class="external-link">ieugwasr</a> package to
connect to the database of thousands of complete GWAS summary data.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install directly from the GitHub repository do the following:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"MRCIEU/TwoSampleMR"</span><span class="op">)</span></span></code></pre></div>
<p>If you don’t have the <code>remotes</code> package install it from
CRAN using <code>install.packages("remotes")</code>.</p>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The workflow for performing MR is as follows:</p>
<ol style="list-style-type: decimal">
<li>Select instruments for the exposure (perform LD clumping if
necessary)</li>
<li>Extract the instruments from the <a href="https://gwas.mrcieu.ac.uk/" class="external-link">IEU GWAS database</a> for the outcomes
of interest</li>
<li>Harmonise the effect sizes for the instruments on the exposures and
the outcomes to be each for the same reference allele</li>
<li>Perform MR analysis, sensitivity analyses, create plots, compile
reports</li>
</ol>
<p>A diagrammatic overview is shown here:</p>
<p><img src="img%2Ftwosamplemr_schematic_long-01.png" alt="A diagrammatic overview of performing a two-sample Mendelian randomization analysis" width="800px" style="display: block; margin: auto;"></p>
<p>A basic analysis, e.g. the causal effect of body mass index on
coronary heart disease, looks like this:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/TwoSampleMR" class="external-link">TwoSampleMR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List available GWASs</span></span>
<span><span class="va">ao</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/available_outcomes.html">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get instruments</span></span>
<span><span class="va">exposure_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_instruments.html">extract_instruments</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get effects of instruments on outcome</span></span>
<span><span class="va">outcome_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span>snps<span class="op">=</span><span class="va">exposure_dat</span><span class="op">$</span><span class="va">SNP</span>, outcomes <span class="op">=</span> <span class="st">"ieu-a-7"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Harmonise the exposure and outcome data</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/harmonise_data.html">harmonise_data</a></span><span class="op">(</span><span class="va">exposure_dat</span>, <span class="va">outcome_dat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform MR</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mr.html">mr</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></code></pre></div>
<p>Each step is documented on other pages in the documentation.</p>
</div>
<div class="section level2">
<h2 id="authentication">Authentication<a class="anchor" aria-label="anchor" href="#authentication"></a>
</h2>
<p>The statistical methods in TwoSampleMR can be used on any data, but
there are a number of functions that connect to the OpenGWAS database
for data extraction. These OpenGWAS data access functions require
authentication.</p>
<p><strong>Authentication is changing</strong> The main differences are
that:</p>
<ol style="list-style-type: decimal">
<li>Authentication is required for most queries to OpenGWAS for everyone
(i.e. no more anonymous usage)</li>
<li>We are no longer using Google Oauth2. This has been replaced by a
simple API key system.</li>
</ol>
<p>Detailed information is given here: <a href="https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication" class="external-link uri">https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication</a>.</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Bowden2015" class="csl-entry">
Bowden, Jack, George Davey Smith, and Stephen Burgess. 2015.
<span>“<span class="nocase">Mendelian randomization with invalid
instruments: effect estimation and bias detection through Egger
regression</span>.”</span> <em>International Journal of
Epidemiology</em> 44 (2): 512–25. <a href="https://doi.org/10.1093/ije/dyv080" class="external-link">https://doi.org/10.1093/ije/dyv080</a>.
</div>
<div id="ref-DaveySmith2003" class="csl-entry">
Davey Smith, G., and S. Ebrahim. 2003. <span>“<span class="nocase">’Mendelian randomization’: can genetic epidemiology
contribute to understanding environmental determinants of
disease?</span>”</span> <em>International Journal of Epidemiology</em>
32 (1): 1–22. <a href="https://doi.org/10.1093/ije/dyg070" class="external-link">https://doi.org/10.1093/ije/dyg070</a>.
</div>
<div id="ref-DaveySmithHemani2014" class="csl-entry">
Davey Smith, George, and Gibran Hemani. 2014. <span>“<span class="nocase">Mendelian randomization: genetic anchors for causal
inference in epidemiological studies</span>.”</span> <em>Human Molecular
Genetics</em> 23 (R1): R89–98. <a href="https://doi.org/10.1093/hmg/ddu328" class="external-link">https://doi.org/10.1093/hmg/ddu328</a>.
</div>
<div id="ref-Pierce2013" class="csl-entry">
Pierce, Brandon L, and Stephen Burgess. 2013. <span>“<span class="nocase">Efficient design for Mendelian randomization studies:
subsample and 2-sample instrumental variable estimators</span>.”</span>
<em>American Journal of Epidemiology</em> 178 (7): 1177–84. <a href="https://doi.org/10.1093/aje/kwt084" class="external-link">https://doi.org/10.1093/aje/kwt084</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
