<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Outcome data • TwoSampleMR</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Outcome data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guide">
<li><a class="dropdown-item" href="../articles/index.html">List of articles</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/exposure.html">Exposure data</a></li>
    <li><a class="dropdown-item" href="../articles/outcome.html">Outcome data</a></li>
    <li><a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a></li>
    <li><a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a></li>
    <li><a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/TwoSampleMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Outcome data</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/master/vignettes/outcome.Rmd" class="external-link"><code>vignettes/outcome.Rmd</code></a></small>
      <div class="d-none name"><code>outcome.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/TwoSampleMR" class="external-link">TwoSampleMR</a></span><span class="op">)</span></span></code></pre></div>
<p>Once instruments for the exposure trait have been specified, those
variants need to be extracted from the outcome trait.</p>
<div class="section level2">
<h2 id="available-studies-in-ieu-gwas-database">Available studies in IEU GWAS database<a class="anchor" aria-label="anchor" href="#available-studies-in-ieu-gwas-database"></a>
</h2>
<p>The IEU GWAS database (IGD) contains complete GWAS summary statistics
from a large number of studies. You can browse them here:</p>
<p><a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a></p>
<p>To obtain details about the available GWASs programmatically do the
following:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/available_outcomes.html">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ao</span><span class="op">)</span></span>
<span><span class="co">#&gt;           id         trait ncase group_name year       author consortium</span></span>
<span><span class="co">#&gt; 1 ieu-b-5103 Schizophrenia  1234     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt; 2 ieu-b-5102 Schizophrenia 52017     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt; 3 ieu-b-5101 Schizophrenia 12305     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt; 4 ieu-b-5100 Schizophrenia 64322     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt; 5 ieu-b-5099 Schizophrenia 76755     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt; 6 ieu-b-5098 Schizophrenia  5998     public 2022 Trubetskoy V        PGC</span></span>
<span><span class="co">#&gt;                 sex     pmid                         population  unit</span></span>
<span><span class="co">#&gt; 1 Males and Females 35396580         Hispanic or Latin American logOR</span></span>
<span><span class="co">#&gt; 2 Males and Females 35396580                           European logOR</span></span>
<span><span class="co">#&gt; 3 Males and Females 35396580                         East Asian logOR</span></span>
<span><span class="co">#&gt; 4 Males and Females 35396580                              Mixed logOR</span></span>
<span><span class="co">#&gt; 5 Males and Females 35396580                              Mixed logOR</span></span>
<span><span class="co">#&gt; 6 Males and Females 35396580 African American or Afro-Caribbean logOR</span></span>
<span><span class="co">#&gt;   sample_size       build ncontrol category subcategory      ontology</span></span>
<span><span class="co">#&gt; 1        4324 HG19/GRCh37     3090  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt; 2      127906 HG19/GRCh37    75889  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt; 3       27363 HG19/GRCh37    15058  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt; 4      155269 HG19/GRCh37    90947  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt; 5      320404 HG19/GRCh37   243649  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt; 6        9824 HG19/GRCh37     3826  Disease          NA MONDO:0005090</span></span>
<span><span class="co">#&gt;                                                                      note mr</span></span>
<span><span class="co">#&gt; 1                                                                    &lt;NA&gt; NA</span></span>
<span><span class="co">#&gt; 2                                                                    &lt;NA&gt; NA</span></span>
<span><span class="co">#&gt; 3                                                                    &lt;NA&gt; NA</span></span>
<span><span class="co">#&gt; 4                            Core - East Asian and European meta analysis NA</span></span>
<span><span class="co">#&gt; 5 Primary - meta analysis of Eur, East Asian, African American and Latino NA</span></span>
<span><span class="co">#&gt; 6                                                                    &lt;NA&gt; NA</span></span>
<span><span class="co">#&gt;   nsnp  doi coverage study_design priority sd</span></span>
<span><span class="co">#&gt; 1   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span>
<span><span class="co">#&gt; 2   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span>
<span><span class="co">#&gt; 3   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span>
<span><span class="co">#&gt; 4   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span>
<span><span class="co">#&gt; 5   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span>
<span><span class="co">#&gt; 6   NA &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;       NA NA</span></span></code></pre></div>
<p>For information about authentication see <a href="https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication" class="external-link uri">https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication</a>.</p>
<p>The <code>available_outcomes</code> function returns a table of all
the available studies in the database. Each study has a unique ID.
e.g.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ao</span>, select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">trait</span>, <span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           trait         id</span></span>
<span><span class="co">#&gt; 1 Schizophrenia ieu-b-5103</span></span>
<span><span class="co">#&gt; 2 Schizophrenia ieu-b-5102</span></span>
<span><span class="co">#&gt; 3 Schizophrenia ieu-b-5101</span></span>
<span><span class="co">#&gt; 4 Schizophrenia ieu-b-5100</span></span>
<span><span class="co">#&gt; 5 Schizophrenia ieu-b-5099</span></span>
<span><span class="co">#&gt; 6 Schizophrenia ieu-b-5098</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-particular-snps-from-particular-studies">Extracting particular SNPs from particular studies<a class="anchor" aria-label="anchor" href="#extracting-particular-snps-from-particular-studies"></a>
</h2>
<p>If we want to perform MR of BMI against coronary heart disease, we
need to identify the SNPs that influence the BMI, and then extract those
SNPs from a GWAS on coronary heart disease.</p>
<p>Let’s get the Locke et al 2014 instruments for BMI as an example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bmi_exp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_instruments.html">extract_instruments</a></span><span class="op">(</span>outcomes <span class="op">=</span> <span class="st">'ieu-a-2'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bmi_exp_dat</span><span class="op">)</span></span>
<span><span class="co">#&gt;   pval.exposure samplesize.exposure chr.exposure se.exposure beta.exposure</span></span>
<span><span class="co">#&gt; 1   2.18198e-08              339152            1      0.0030       -0.0168</span></span>
<span><span class="co">#&gt; 2   4.56773e-11              339065            1      0.0031        0.0201</span></span>
<span><span class="co">#&gt; 3   5.05941e-14              313621            1      0.0087        0.0659</span></span>
<span><span class="co">#&gt; 4   5.45205e-10              338768            1      0.0029        0.0181</span></span>
<span><span class="co">#&gt; 5   1.88018e-28              338123            1      0.0030        0.0331</span></span>
<span><span class="co">#&gt; 6   2.28718e-40              339078            1      0.0037        0.0497</span></span>
<span><span class="co">#&gt;   pos.exposure id.exposure        SNP effect_allele.exposure</span></span>
<span><span class="co">#&gt; 1     47684677     ieu-a-2   rs977747                      G</span></span>
<span><span class="co">#&gt; 2     78048331     ieu-a-2 rs17381664                      C</span></span>
<span><span class="co">#&gt; 3    110082886     ieu-a-2  rs7550711                      T</span></span>
<span><span class="co">#&gt; 4    201784287     ieu-a-2  rs2820292                      C</span></span>
<span><span class="co">#&gt; 5     72837239     ieu-a-2  rs7531118                      C</span></span>
<span><span class="co">#&gt; 6    177889480     ieu-a-2   rs543874                      G</span></span>
<span><span class="co">#&gt;   other_allele.exposure eaf.exposure                      exposure</span></span>
<span><span class="co">#&gt; 1                     T       0.5333 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 2                     T       0.4250 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 3                     C       0.0339 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 4                     A       0.5083 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 5                     T       0.6083 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 6                     A       0.2667 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt;   mr_keep.exposure pval_origin.exposure data_source.exposure</span></span>
<span><span class="co">#&gt; 1             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 2             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 3             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 4             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 5             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 6             TRUE             reported                  igd</span></span></code></pre></div>
<p>We now need to find a suitable GWAS for coronary heart disease. We
can search the available studies:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"heart disease"</span>, <span class="va">ao</span><span class="op">$</span><span class="va">trait</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;                                 id</span></span>
<span><span class="co">#&gt; 7958                 finn-b-I9_CHD</span></span>
<span><span class="co">#&gt; 10829                   ukb-b-3983</span></span>
<span><span class="co">#&gt; 12268                ukb-e-I25_AFR</span></span>
<span><span class="co">#&gt; 14221                   ukb-b-2205</span></span>
<span><span class="co">#&gt; 14897                      ieu-a-7</span></span>
<span><span class="co">#&gt; 15028 finn-b-I9_SECONDRIGHT_EXNONE</span></span>
<span><span class="co">#&gt; 16883                   ukb-b-7436</span></span>
<span><span class="co">#&gt; 18046                    ukb-a-534</span></span>
<span><span class="co">#&gt; 18048                ukb-e-I25_CSA</span></span>
<span><span class="co">#&gt; 20207           finn-b-I9_OTHHEART</span></span>
<span><span class="co">#&gt; 22249         finn-b-I9_VHD_EXNONE</span></span>
<span><span class="co">#&gt; 22725          finn-b-I9_PULMHEART</span></span>
<span><span class="co">#&gt; 23614                      ieu-a-9</span></span>
<span><span class="co">#&gt; 24603           finn-b-FG_OTHHEART</span></span>
<span><span class="co">#&gt; 27414             ebi-a-GCST000998</span></span>
<span><span class="co">#&gt; 33264          finn-b-FG_PULMHEART</span></span>
<span><span class="co">#&gt; 35220                 ukb-d-I9_CHD</span></span>
<span><span class="co">#&gt; 35231        finn-b-I9_OTHILLHEART</span></span>
<span><span class="co">#&gt; 36931                   ukb-b-8184</span></span>
<span><span class="co">#&gt; 37302 finn-b-I9_OTHILLHEART_EXNONE</span></span>
<span><span class="co">#&gt; 37894                   ukb-b-1668</span></span>
<span><span class="co">#&gt; 38319                finn-b-I9_IHD</span></span>
<span><span class="co">#&gt; 38448            finn-b-I9_RHEUFEV</span></span>
<span><span class="co">#&gt; 38602                      ieu-a-8</span></span>
<span><span class="co">#&gt; 41994          finn-b-I9_ISCHHEART</span></span>
<span><span class="co">#&gt; 42095                 ukb-d-I9_IHD</span></span>
<span><span class="co">#&gt; 42882        finn-b-I9_SECONDRIGHT</span></span>
<span><span class="co">#&gt; 43747           ukb-d-I9_CHD_NOREV</span></span>
<span><span class="co">#&gt; 45294                      ieu-a-6</span></span>
<span><span class="co">#&gt; 46126                finn-b-I9_VHD</span></span>
<span><span class="co">#&gt; 46689                  ukb-b-16606</span></span>
<span><span class="co">#&gt;                                                                                                                           trait</span></span>
<span><span class="co">#&gt; 7958                                                                                         Major coronary heart disease event</span></span>
<span><span class="co">#&gt; 10829                                                Diagnoses - main ICD10: I25.9 Chronic ischaemic heart disease, unspecified</span></span>
<span><span class="co">#&gt; 12268                                                                                       I25 Chronic ischaemic heart disease</span></span>
<span><span class="co">#&gt; 14221 Diagnoses - secondary ICD10: Z82.4 Family history of ischaemic heart disease and other diseases of the circulatory system</span></span>
<span><span class="co">#&gt; 14897                                                                                                    Coronary heart disease</span></span>
<span><span class="co">#&gt; 15028                                                                      Secondary right heart disease (no controls excluded)</span></span>
<span><span class="co">#&gt; 16883                                                          Diagnoses - secondary ICD10: I25.1 Atherosclerotic heart disease</span></span>
<span><span class="co">#&gt; 18046                                                               Diagnoses - main ICD10: I25 Chronic ischaemic heart disease</span></span>
<span><span class="co">#&gt; 18048                                                                                       I25 Chronic ischaemic heart disease</span></span>
<span><span class="co">#&gt; 20207                                                                                        Other heart diseases (I9_OTHHEART)</span></span>
<span><span class="co">#&gt; 22249                                                   Valvular heart disease including rheumatic fever (no controls excluded)</span></span>
<span><span class="co">#&gt; 22725                                                                Pulmonary heart disease, diseases of pulmonary circulation</span></span>
<span><span class="co">#&gt; 23614                                                                                                    Coronary heart disease</span></span>
<span><span class="co">#&gt; 24603                                                                                        Other heart diseases (FG_OTHHEART)</span></span>
<span><span class="co">#&gt; 27414                                                                                                    Coronary heart disease</span></span>
<span><span class="co">#&gt; 33264                                                                                                   Pulmonary heart disease</span></span>
<span><span class="co">#&gt; 35220                                                                                        Major coronary heart disease event</span></span>
<span><span class="co">#&gt; 35231                                                                                       Other or ill-defined heart diseases</span></span>
<span><span class="co">#&gt; 36931                                           Diagnoses - secondary ICD10: I25.9 Chronic ischaemic heart disease, unspecified</span></span>
<span><span class="co">#&gt; 37302                                                                Other or ill-defined heart diseases (no controls excluded)</span></span>
<span><span class="co">#&gt; 37894                                                               Diagnoses - main ICD10: I25.1 Atherosclerotic heart disease</span></span>
<span><span class="co">#&gt; 38319                                                                                  Ischaemic heart disease, wide definition</span></span>
<span><span class="co">#&gt; 38448                                                                                        Rheumatic fever incl heart disease</span></span>
<span><span class="co">#&gt; 38602                                                                                                    Coronary heart disease</span></span>
<span><span class="co">#&gt; 41994                                                                                                   Ischemic heart diseases</span></span>
<span><span class="co">#&gt; 42095                                                                                  Ischaemic heart disease, wide definition</span></span>
<span><span class="co">#&gt; 42882                                                                                             Secondary right heart disease</span></span>
<span><span class="co">#&gt; 43747                                                           Major coronary heart disease event excluding revascularizations</span></span>
<span><span class="co">#&gt; 45294                                                                                                    Coronary heart disease</span></span>
<span><span class="co">#&gt; 46126                                                                          Valvular heart disease including rheumatic fever</span></span>
<span><span class="co">#&gt; 46689                                         Diagnoses - secondary ICD10: I25.8 Other forms of chronic ischaemic heart disease</span></span>
<span><span class="co">#&gt;       ncase group_name year       author        consortium               sex</span></span>
<span><span class="co">#&gt; 7958  21012     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 10829  1195     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt; 12268   302     public 2020 Pan-UKB team                NA Males and Females</span></span>
<span><span class="co">#&gt; 14221  9330     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt; 14897 60801     public 2015       Nikpay CARDIoGRAMplusC4D Males and Females</span></span>
<span><span class="co">#&gt; 15028   428     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 16883  5771     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt; 18046  8755     public 2017        Neale         Neale Lab Males and Females</span></span>
<span><span class="co">#&gt; 18048  1205     public 2020 Pan-UKB team                NA Males and Females</span></span>
<span><span class="co">#&gt; 20207 62081     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 22249 38209     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 22725  4564     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 23614 63746     public 2013     Deloukas CARDIoGRAMplusC4D Males and Females</span></span>
<span><span class="co">#&gt; 24603 58173     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 27414 22233     public 2011  Schunkert H                NA                NA</span></span>
<span><span class="co">#&gt; 33264  4185     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 35220 10157     public 2018    Neale lab                NA Males and Females</span></span>
<span><span class="co">#&gt; 35231   713     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 36931  5861     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt; 37302   713     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 37894 12171     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt; 38319 31640     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 38448   573     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 38602 22233     public 2011  Schunkert H        CARDIoGRAM Males and Females</span></span>
<span><span class="co">#&gt; 41994 30952     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 42095 20857     public 2018    Neale lab                NA Males and Females</span></span>
<span><span class="co">#&gt; 42882   428     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 43747 10157     public 2018    Neale lab                NA Males and Females</span></span>
<span><span class="co">#&gt; 45294 15420     public 2011        Peden               C4D Males and Females</span></span>
<span><span class="co">#&gt; 46126 38209     public 2021           NA                NA Males and Females</span></span>
<span><span class="co">#&gt; 46689  5738     public 2018 Ben Elsworth           MRC-IEU Males and Females</span></span>
<span><span class="co">#&gt;           pmid                         population     unit sample_size</span></span>
<span><span class="co">#&gt; 7958        NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 10829       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt; 12268       NA African American or Afro-Caribbean       NA        6636</span></span>
<span><span class="co">#&gt; 14221       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt; 14897 26343387                              Mixed log odds      184305</span></span>
<span><span class="co">#&gt; 15028       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 16883       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt; 18046       NA                           European       SD      337199</span></span>
<span><span class="co">#&gt; 18048       NA                        South Asian       NA        8876</span></span>
<span><span class="co">#&gt; 20207       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 22249       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 22725       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 23614 23202125                              Mixed log odds      194427</span></span>
<span><span class="co">#&gt; 24603       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 27414 21378990                           European    logOR       86995</span></span>
<span><span class="co">#&gt; 33264       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 35220       NA                           European       NA      361194</span></span>
<span><span class="co">#&gt; 35231       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 36931       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt; 37302       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 37894       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt; 38319       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 38448       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 38602 21378990                           European log odds       86995</span></span>
<span><span class="co">#&gt; 41994       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 42095       NA                           European       NA      361194</span></span>
<span><span class="co">#&gt; 42882       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 43747       NA                           European       NA      361194</span></span>
<span><span class="co">#&gt; 45294 21378988                              Mixed log odds       30482</span></span>
<span><span class="co">#&gt; 46126       NA                           European       NA          NA</span></span>
<span><span class="co">#&gt; 46689       NA                           European       SD      463010</span></span>
<span><span class="co">#&gt;             build ncontrol category    subcategory ontology</span></span>
<span><span class="co">#&gt; 7958  HG19/GRCh37   197780   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 10829 HG19/GRCh37   461815   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 12268 HG19/GRCh37     6334   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 14221 HG19/GRCh37   453680   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 14897 HG19/GRCh37   123504  Disease Cardiovascular       NA</span></span>
<span><span class="co">#&gt; 15028 HG19/GRCh37   218364   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 16883 HG19/GRCh37   457239   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 18046 HG19/GRCh37   328444       NA             NA       NA</span></span>
<span><span class="co">#&gt; 18048 HG19/GRCh37     7671   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 20207 HG19/GRCh37   156711   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 22249 HG19/GRCh37   180583   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 22725 HG19/GRCh37   214228   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 23614 HG19/GRCh37   130681  Disease Cardiovascular       NA</span></span>
<span><span class="co">#&gt; 24603 HG19/GRCh37   160619   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 27414 HG19/GRCh37    64762       NA             NA       NA</span></span>
<span><span class="co">#&gt; 33264 HG19/GRCh37   214607   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 35220 HG19/GRCh37   351037   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 35231 HG19/GRCh37   156711   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 36931 HG19/GRCh37   457149   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 37302 HG19/GRCh37   218079   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 37894 HG19/GRCh37   450839   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 38319 HG19/GRCh37   187152   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 38448 HG19/GRCh37   218219   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 38602 HG19/GRCh37    64762  Disease Cardiovascular       NA</span></span>
<span><span class="co">#&gt; 41994 HG19/GRCh37   187840   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 42095 HG19/GRCh37   340337   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 42882 HG19/GRCh37   214228   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 43747 HG19/GRCh37   351037   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 45294 HG19/GRCh37    15062  Disease Cardiovascular       NA</span></span>
<span><span class="co">#&gt; 46126 HG19/GRCh37   156711   Binary             NA       NA</span></span>
<span><span class="co">#&gt; 46689 HG19/GRCh37   457272   Binary             NA       NA</span></span>
<span><span class="co">#&gt;                                                                                       note</span></span>
<span><span class="co">#&gt; 7958                                                                                I9_CHD</span></span>
<span><span class="co">#&gt; 10829 41202#I259: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt; 12268                                                                                   NA</span></span>
<span><span class="co">#&gt; 14221 41204#Z824: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt; 14897                                                                                 &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 15028                                                                I9_SECONDRIGHT_EXNONE</span></span>
<span><span class="co">#&gt; 16883 41204#I251: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt; 18046                                                                                   NA</span></span>
<span><span class="co">#&gt; 18048                                                                                   NA</span></span>
<span><span class="co">#&gt; 20207                                                                          I9_OTHHEART</span></span>
<span><span class="co">#&gt; 22249                                                                        I9_VHD_EXNONE</span></span>
<span><span class="co">#&gt; 22725                                                                         I9_PULMHEART</span></span>
<span><span class="co">#&gt; 23614                                                                                 &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 24603                                                                          FG_OTHHEART</span></span>
<span><span class="co">#&gt; 27414                                                                                   NA</span></span>
<span><span class="co">#&gt; 33264                                                                         FG_PULMHEART</span></span>
<span><span class="co">#&gt; 35220                                                                                   NA</span></span>
<span><span class="co">#&gt; 35231                                                                       I9_OTHILLHEART</span></span>
<span><span class="co">#&gt; 36931 41204#I259: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt; 37302                                                                I9_OTHILLHEART_EXNONE</span></span>
<span><span class="co">#&gt; 37894 41202#I251: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt; 38319                                                                               I9_IHD</span></span>
<span><span class="co">#&gt; 38448                                                                           I9_RHEUFEV</span></span>
<span><span class="co">#&gt; 38602                                                                                 &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 41994                                                                         I9_ISCHHEART</span></span>
<span><span class="co">#&gt; 42095                                                                                   NA</span></span>
<span><span class="co">#&gt; 42882                                                                       I9_SECONDRIGHT</span></span>
<span><span class="co">#&gt; 43747                                                                                   NA</span></span>
<span><span class="co">#&gt; 45294                                                                                 &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 46126                                                                               I9_VHD</span></span>
<span><span class="co">#&gt; 46689 41204#I258: Output from GWAS pipeline using Phesant derived variables from UKBiobank</span></span>
<span><span class="co">#&gt;       mr     nsnp  doi coverage study_design priority sd</span></span>
<span><span class="co">#&gt; 7958   1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 10829  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 12268  1 15478580 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 14221  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 14897  1  9455779 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 15028  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 16883  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 18046  1 10894596 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 18048  1  9811287 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 20207  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 22249  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 22725  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 23614  1    79129 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 24603  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 27414  1  2415020 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 33264  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 35220  1 13295130 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 35231  1 16380177 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 36931  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 37302  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 37894  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span>
<span><span class="co">#&gt; 38319  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 38448  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 38602  1  2420361 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        2 NA</span></span>
<span><span class="co">#&gt; 41994  1 16380466 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 42095  1 13586589 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 42882  1 16380459 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 43747  1 13295130 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 45294  1   540233 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        3 NA</span></span>
<span><span class="co">#&gt; 46126  1 16380358 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        0 NA</span></span>
<span><span class="co">#&gt; 46689  1  9851867 &lt;NA&gt;     &lt;NA&gt;         &lt;NA&gt;        1 NA</span></span></code></pre></div>
<p>The most recent CARDIOGRAM GWAS is ID number <code>ieu-a-7</code>. We
can extract the BMI SNPs from this GWAS as follows:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chd_out_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span></span>
<span>    snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>,</span>
<span>    outcomes <span class="op">=</span> <span class="st">'ieu-a-7'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/extract_outcome_data.html">extract_outcome_data()</a></code> function is flexible. The
<code>snps</code> argument only requires an array of rsIDs, and the
<code>outcomes</code> argument can be a vector of outcomes, e.g.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chd_out_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span></span>
<span>    snps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs234"</span>, <span class="st">"rs17097147"</span><span class="op">)</span>,</span>
<span>    outcomes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ieu-a-2'</span>, <span class="st">'ieu-a-7'</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>will extract the two SNPs from each of the outcomes
<code>ieu-a-2</code> and <code>ieu-a-7</code>.</p>
</div>
<div class="section level2">
<h2 id="ld-proxies">LD proxies<a class="anchor" aria-label="anchor" href="#ld-proxies"></a>
</h2>
<p>By default if a particular requested SNP is not present in the
outcome GWAS then a SNP (proxy) that is in LD with the requested SNP
(target) will be searched for instead. LD proxies are defined using 1000
genomes European sample data. The effect of the proxy SNP on the outcome
is returned, along with the proxy SNP, the effect allele of the proxy
SNP, and the corresponding allele (in phase) for the target SNP.</p>
<p>The parameters for handling LD proxies are as follows:</p>
<ul>
<li>
<code>proxies</code> = TRUE or FALSE (TRUE by default)</li>
<li>
<code>rsq</code> = numeric value of minimum rsq to find a proxy.
Default is 0.8, minimum is 0.6</li>
<li>
<code>palindromes</code> = Allow palindromic SNPs? Default is 1
(yes)</li>
<li>
<code>maf_threshold</code> = If palindromes allowed then what is the
maximum minor allele frequency of palindromes allowed? Default is
0.3.</li>
</ul>
</div>
<div class="section level2">
<h2 id="using-local-gwas-summary-data">Using local GWAS summary data<a class="anchor" aria-label="anchor" href="#using-local-gwas-summary-data"></a>
</h2>
<p>If you have GWAS summary data that is not present in IEU GWAS
database, this can still be used to perform analysis.</p>
<p>Supposing there is a GWAS summary file called “gwas_summary.csv” with
e.g. 2 million rows and it looks like this:</p>
<pre><code>rsid,effect,SE,a1,a2,a1_freq,p-value,Units,Gene,n
rs10767664,0.19,0.030612245,A,T,0.78,5.00E-26,kg/m2,BDNF,225238
rs13078807,0.1,0.020408163,G,A,0.2,4.00E-11,kg/m2,CADM2,221431
rs1514175,0.07,0.020408163,A,G,0.43,8.00E-14,kg/m2,TNNI3K,207641
rs1558902,0.39,0.020408163,A,T,0.42,5.00E-120,kg/m2,FTO,222476
...
...</code></pre>
<p>To extract the exposure SNPs from this data, we would use the
following command:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outcome_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_outcome_data.html">read_outcome_data</a></span><span class="op">(</span></span>
<span>    snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>,</span>
<span>    filename <span class="op">=</span> <span class="st">"gwas_summary.csv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">","</span>,</span>
<span>    snp_col <span class="op">=</span> <span class="st">"rsid"</span>,</span>
<span>    beta_col <span class="op">=</span> <span class="st">"effect"</span>,</span>
<span>    se_col <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>    effect_allele_col <span class="op">=</span> <span class="st">"a1"</span>,</span>
<span>    other_allele_col <span class="op">=</span> <span class="st">"a2"</span>,</span>
<span>    eaf_col <span class="op">=</span> <span class="st">"a1_freq"</span>,</span>
<span>    pval_col <span class="op">=</span> <span class="st">"p-value"</span>,</span>
<span>    units_col <span class="op">=</span> <span class="st">"Units"</span>,</span>
<span>    gene_col <span class="op">=</span> <span class="st">"Gene"</span>,</span>
<span>    samplesize_col <span class="op">=</span> <span class="st">"n"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This returns an outcome data frame with only the SNPs that were
requested (if those SNPs were present in the “gwas_summary.csv”
file).</p>
</div>
<div class="section level2">
<h2 id="outcome-data-format">Outcome data format<a class="anchor" aria-label="anchor" href="#outcome-data-format"></a>
</h2>
<p>The <code>extract_outcome_data</code> function returns a table of SNP
effects for the requested SNPs on the requested outcomes. The format of
the data is similar to the exposure data format, except the main columns
are as follows:</p>
<ul>
<li><code>SNP</code></li>
<li><code>beta.outcome</code></li>
<li><code>se.outcome</code></li>
<li><code>samplesize.outcome</code></li>
<li><code>ncase.outcome</code></li>
<li><code>ncontrol.outcome</code></li>
<li><code>pval.outcome</code></li>
<li><code>eaf.outcome</code></li>
<li><code>effect_allele.outcom</code></li>
<li><code>other_allele.outcome</code></li>
<li><code>units.outcome</code></li>
<li><code>outcome</code></li>
<li><code>consortium.outcome</code></li>
<li><code>year.outcome</code></li>
<li><code>pmid.outcome</code></li>
<li><code>id.outcome</code></li>
<li><code>originalname.outcome</code></li>
<li><code>proxy.outcome</code></li>
<li><code>target_snp.outcome</code></li>
<li><code>proxy_snp.outcome</code></li>
<li><code>target_a1.outcome</code></li>
<li><code>target_a2.outcome</code></li>
<li><code>proxy_a1.outcome</code></li>
<li><code>proxy_a2.outcome</code></li>
<li><code>mr_keep.outcome</code></li>
<li><code>data_source.outcome</code></li>
</ul>
</div>
<div class="section level2">
<h2 id="more-advanced-use-of-local-data">More advanced use of local data<a class="anchor" aria-label="anchor" href="#more-advanced-use-of-local-data"></a>
</h2>
<p>We have developed a summary data format called “GWAS VCF”, which is
designed to store GWAS results in a strict and performant way. It is
possible to use this format with the TwoSampleMR package. Going down
this avenue also allows you to use LD proxy functionality using your own
LD reference files (or ones that we provide). For more details, see this
package that explains the format and how to query it in R:</p>
<p><a href="https://github.com/mrcieu/gwasvcf" class="external-link uri">https://github.com/mrcieu/gwasvcf</a></p>
<p>and this package for how to connect the data to other packages
including TwoSampleMR</p>
<p><a href="https://github.com/MRCIEU/gwasglue" class="external-link uri">https://github.com/MRCIEU/gwasglue</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
