<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="TwoSampleMR">
<title>Outcome data • TwoSampleMR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Outcome data">
<meta property="og:description" content="TwoSampleMR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TwoSampleMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-guide">Guide</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-guide">
    <a class="dropdown-item" href="../articles/index.html">List of articles</a>
    <a class="dropdown-item" href="../articles/introduction.html">Introduction</a>
    <a class="dropdown-item" href="../articles/exposure.html">Exposure data</a>
    <a class="dropdown-item" href="../articles/outcome.html">Outcome data</a>
    <a class="dropdown-item" href="../articles/harmonise.html">Harmonise data</a>
    <a class="dropdown-item" href="../articles/perform_mr.html">Perform MR</a>
    <a class="dropdown-item" href="../articles/gwas2020.html">Major changes to the IEU GWAS resource for 2020</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrcieu/TwoSampleMR">Source</a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Outcome data</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/TwoSampleMR/blob/HEAD/vignettes/outcome.Rmd" class="external-link"><code>vignettes/outcome.Rmd</code></a></small>
      <div class="d-none name"><code>outcome.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/TwoSampleMR" class="external-link">TwoSampleMR</a></span><span class="op">)</span></span></code></pre></div>
<p>Once instruments for the exposure trait have been specified, those
variants need to be extracted from the outcome trait.</p>
<div class="section level2">
<h2 id="available-studies-in-ieu-gwas-database">Available studies in IEU GWAS database<a class="anchor" aria-label="anchor" href="#available-studies-in-ieu-gwas-database"></a>
</h2>
<p>The IEU GWAS database (IGD) contains complete GWAS summary statistics
from a large number of studies. You can browse them here:</p>
<p><a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a></p>
<p>To obtain details about the available GWASs programmatically do the
following:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/available_outcomes.html">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; API: public: http://gwas-api.mrcieu.ac.uk/</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ao</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 25</span></span></span>
<span><span class="co">#&gt;   id      trait ncase group_name  year author consortium sex     pmid population</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ieu-b-… Schi…  <span style="text-decoration: underline;">1</span>234 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 Hispanic …</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ieu-b-… Schi… <span style="text-decoration: underline;">52</span>017 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ieu-b-… Schi… <span style="text-decoration: underline;">12</span>305 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 East Asian</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ieu-b-… Schi… <span style="text-decoration: underline;">64</span>322 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 Mixed     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ieu-b-… Schi… <span style="text-decoration: underline;">76</span>755 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 Mixed     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ieu-b-… Schi…  <span style="text-decoration: underline;">5</span>998 public      <span style="text-decoration: underline;">2</span>022 Trube… PGC        Male… 3.54<span style="color: #949494;">e</span>7 African A…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 15 more variables: unit &lt;chr&gt;, sample_size &lt;int&gt;, build &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ncontrol &lt;int&gt;, category &lt;chr&gt;, subcategory &lt;chr&gt;, ontology &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   note &lt;chr&gt;, mr &lt;int&gt;, nsnp &lt;int&gt;, doi &lt;chr&gt;, coverage &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   study_design &lt;chr&gt;, priority &lt;int&gt;, sd &lt;dbl&gt;</span></span></span></code></pre></div>
<p>For information about authentication see <a href="https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication" class="external-link uri">https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication</a>.</p>
<p>The <code>available_outcomes</code> function returns a table of all
the available studies in the database. Each study has a unique ID.
e.g.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ao</span>, select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">trait</span>, <span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   trait         id        </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Schizophrenia ieu-b-5103</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Schizophrenia ieu-b-5102</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Schizophrenia ieu-b-5101</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Schizophrenia ieu-b-5100</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Schizophrenia ieu-b-5099</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Schizophrenia ieu-b-5098</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-particular-snps-from-particular-studies">Extracting particular SNPs from particular studies<a class="anchor" aria-label="anchor" href="#extracting-particular-snps-from-particular-studies"></a>
</h2>
<p>If we want to perform MR of BMI against coronary heart disease, we
need to identify the SNPs that influence the BMI, and then extract those
SNPs from a GWAS on coronary heart disease.</p>
<p>Let’s get the Locke et al 2014 instruments for BMI as an example:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bmi_exp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_instruments.html">extract_instruments</a></span><span class="op">(</span>outcomes <span class="op">=</span> <span class="st">'ieu-a-2'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bmi_exp_dat</span><span class="op">)</span></span>
<span><span class="co">#&gt;   pval.exposure samplesize.exposure chr.exposure se.exposure beta.exposure</span></span>
<span><span class="co">#&gt; 1   2.18198e-08              339152            1      0.0030       -0.0168</span></span>
<span><span class="co">#&gt; 2   4.56773e-11              339065            1      0.0031        0.0201</span></span>
<span><span class="co">#&gt; 3   5.05941e-14              313621            1      0.0087        0.0659</span></span>
<span><span class="co">#&gt; 4   5.45205e-10              338768            1      0.0029        0.0181</span></span>
<span><span class="co">#&gt; 5   1.88018e-28              338123            1      0.0030        0.0331</span></span>
<span><span class="co">#&gt; 6   2.28718e-40              339078            1      0.0037        0.0497</span></span>
<span><span class="co">#&gt;   pos.exposure id.exposure        SNP effect_allele.exposure</span></span>
<span><span class="co">#&gt; 1     47684677     ieu-a-2   rs977747                      G</span></span>
<span><span class="co">#&gt; 2     78048331     ieu-a-2 rs17381664                      C</span></span>
<span><span class="co">#&gt; 3    110082886     ieu-a-2  rs7550711                      T</span></span>
<span><span class="co">#&gt; 4    201784287     ieu-a-2  rs2820292                      C</span></span>
<span><span class="co">#&gt; 5     72837239     ieu-a-2  rs7531118                      C</span></span>
<span><span class="co">#&gt; 6    177889480     ieu-a-2   rs543874                      G</span></span>
<span><span class="co">#&gt;   other_allele.exposure eaf.exposure                      exposure</span></span>
<span><span class="co">#&gt; 1                     T       0.5333 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 2                     T       0.4250 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 3                     C       0.0339 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 4                     A       0.5083 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 5                     T       0.6083 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt; 6                     A       0.2667 Body mass index || id:ieu-a-2</span></span>
<span><span class="co">#&gt;   mr_keep.exposure pval_origin.exposure data_source.exposure</span></span>
<span><span class="co">#&gt; 1             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 2             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 3             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 4             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 5             TRUE             reported                  igd</span></span>
<span><span class="co">#&gt; 6             TRUE             reported                  igd</span></span></code></pre></div>
<p>We now need to find a suitable GWAS for coronary heart disease. We
can search the available studies:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ao</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"heart disease"</span>, <span class="va">ao</span><span class="op">$</span><span class="va">trait</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 31 × 25</span></span></span>
<span><span class="co">#&gt;    id    trait ncase group_name  year author consortium sex      pmid population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> finn… Majo… <span style="text-decoration: underline;">21</span>012 public      <span style="text-decoration: underline;">2</span>021 NA     NA         Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ukb-… Diag…  <span style="text-decoration: underline;">1</span>195 public      <span style="text-decoration: underline;">2</span>018 Ben E… MRC-IEU    Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ukb-… I25 …   302 public      <span style="text-decoration: underline;">2</span>020 Pan-U… NA         Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  African A…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ukb-… Diag…  <span style="text-decoration: underline;">9</span>330 public      <span style="text-decoration: underline;">2</span>018 Ben E… MRC-IEU    Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ieu-… Coro… <span style="text-decoration: underline;">60</span>801 public      <span style="text-decoration: underline;">2</span>015 Nikpay CARDIoGRA… Male…  2.63<span style="color: #949494;">e</span>7 Mixed     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> finn… Seco…   428 public      <span style="text-decoration: underline;">2</span>021 NA     NA         Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ukb-… Diag…  <span style="text-decoration: underline;">5</span>771 public      <span style="text-decoration: underline;">2</span>018 Ben E… MRC-IEU    Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ukb-… Diag…  <span style="text-decoration: underline;">8</span>755 public      <span style="text-decoration: underline;">2</span>017 Neale  Neale Lab  Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ukb-… I25 …  <span style="text-decoration: underline;">1</span>205 public      <span style="text-decoration: underline;">2</span>020 Pan-U… NA         Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  South Asi…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> finn… Othe… <span style="text-decoration: underline;">62</span>081 public      <span style="text-decoration: underline;">2</span>021 NA     NA         Male… <span style="color: #BB0000;">NA</span>   <span style="color: #949494;"> </span>  European  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 21 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 15 more variables: unit &lt;chr&gt;, sample_size &lt;int&gt;, build &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ncontrol &lt;int&gt;, category &lt;chr&gt;, subcategory &lt;chr&gt;, ontology &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   note &lt;chr&gt;, mr &lt;int&gt;, nsnp &lt;int&gt;, doi &lt;chr&gt;, coverage &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   study_design &lt;chr&gt;, priority &lt;int&gt;, sd &lt;dbl&gt;</span></span></span></code></pre></div>
<p>The most recent CARDIOGRAM GWAS is ID number <code>ieu-a-7</code>. We
can extract the BMI SNPs from this GWAS as follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chd_out_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span></span>
<span>    snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>,</span>
<span>    outcomes <span class="op">=</span> <span class="st">'ieu-a-7'</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Extracting data for 79 SNP(s) from 1 GWAS(s)</span></span></code></pre></div>
<p>The <code><a href="../reference/extract_outcome_data.html">extract_outcome_data()</a></code> function is flexible. The
<code>snps</code> argument only requires an array of rsIDs, and the
<code>outcomes</code> argument can be a vector of outcomes, e.g.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chd_out_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_outcome_data.html">extract_outcome_data</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs234"</span>, <span class="st">"rs17097147"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ieu-a-2'</span>, <span class="st">'ieu-a-7'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>will extract the two SNPs from each of the outcomes
<code>ieu-a-2</code> and <code>ieu-a-7</code>.</p>
</div>
<div class="section level2">
<h2 id="ld-proxies">LD proxies<a class="anchor" aria-label="anchor" href="#ld-proxies"></a>
</h2>
<p>By default if a particular requested SNP is not present in the
outcome GWAS then a SNP (proxy) that is in LD with the requested SNP
(target) will be searched for instead. LD proxies are defined using 1000
genomes European sample data. The effect of the proxy SNP on the outcome
is returned, along with the proxy SNP, the effect allele of the proxy
SNP, and the corresponding allele (in phase) for the target SNP.</p>
<p>The parameters for handling LD proxies are as follows:</p>
<ul>
<li>
<code>proxies</code> = TRUE or FALSE (TRUE by default)</li>
<li>
<code>rsq</code> = numeric value of minimum rsq to find a proxy.
Default is 0.8, minimum is 0.6</li>
<li>
<code>palindromes</code> = Allow palindromic SNPs? Default is 1
(yes)</li>
<li>
<code>maf_threshold</code> = If palindromes allowed then what is the
maximum minor allele frequency of palindromes allowed? Default is
0.3.</li>
</ul>
</div>
<div class="section level2">
<h2 id="using-local-gwas-summary-data">Using local GWAS summary data<a class="anchor" aria-label="anchor" href="#using-local-gwas-summary-data"></a>
</h2>
<p>If you have GWAS summary data that is not present in IEU GWAS
database, this can still be used to perform analysis.</p>
<p>Supposing there is a GWAS summary file called “gwas_summary.csv” with
e.g. 2 million rows and it looks like this:</p>
<pre><code>rsid,effect,SE,a1,a2,a1_freq,p-value,Units,Gene,n
rs10767664,0.19,0.030612245,A,T,0.78,5.00E-26,kg/m2,BDNF,225238
rs13078807,0.1,0.020408163,G,A,0.2,4.00E-11,kg/m2,CADM2,221431
rs1514175,0.07,0.020408163,A,G,0.43,8.00E-14,kg/m2,TNNI3K,207641
rs1558902,0.39,0.020408163,A,T,0.42,5.00E-120,kg/m2,FTO,222476
...
...</code></pre>
<p>To extract the exposure SNPs from this data, we would use the
following command:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outcome_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_outcome_data.html">read_outcome_data</a></span><span class="op">(</span></span>
<span>    snps <span class="op">=</span> <span class="va">bmi_exp_dat</span><span class="op">$</span><span class="va">SNP</span>,</span>
<span>    filename <span class="op">=</span> <span class="st">"gwas_summary.csv"</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">","</span>,</span>
<span>    snp_col <span class="op">=</span> <span class="st">"rsid"</span>,</span>
<span>    beta_col <span class="op">=</span> <span class="st">"effect"</span>,</span>
<span>    se_col <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>    effect_allele_col <span class="op">=</span> <span class="st">"a1"</span>,</span>
<span>    other_allele_col <span class="op">=</span> <span class="st">"a2"</span>,</span>
<span>    eaf_col <span class="op">=</span> <span class="st">"a1_freq"</span>,</span>
<span>    pval_col <span class="op">=</span> <span class="st">"p-value"</span>,</span>
<span>    units_col <span class="op">=</span> <span class="st">"Units"</span>,</span>
<span>    gene_col <span class="op">=</span> <span class="st">"Gene"</span>,</span>
<span>    samplesize_col <span class="op">=</span> <span class="st">"n"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This returns an outcome data frame with only the SNPs that were
requested (if those SNPs were present in the “gwas_summary.csv”
file).</p>
</div>
<div class="section level2">
<h2 id="outcome-data-format">Outcome data format<a class="anchor" aria-label="anchor" href="#outcome-data-format"></a>
</h2>
<p>The <code>extract_outcome_data</code> function returns a table of SNP
effects for the requested SNPs on the requested outcomes. The format of
the data is similar to the exposure data format, except the main columns
are as follows:</p>
<ul>
<li><code>SNP</code></li>
<li><code>beta.outcome</code></li>
<li><code>se.outcome</code></li>
<li><code>samplesize.outcome</code></li>
<li><code>ncase.outcome</code></li>
<li><code>ncontrol.outcome</code></li>
<li><code>pval.outcome</code></li>
<li><code>eaf.outcome</code></li>
<li><code>effect_allele.outcom</code></li>
<li><code>other_allele.outcome</code></li>
<li><code>units.outcome</code></li>
<li><code>outcome</code></li>
<li><code>consortium.outcome</code></li>
<li><code>year.outcome</code></li>
<li><code>pmid.outcome</code></li>
<li><code>id.outcome</code></li>
<li><code>originalname.outcome</code></li>
<li><code>proxy.outcome</code></li>
<li><code>target_snp.outcome</code></li>
<li><code>proxy_snp.outcome</code></li>
<li><code>target_a1.outcome</code></li>
<li><code>target_a2.outcome</code></li>
<li><code>proxy_a1.outcome</code></li>
<li><code>proxy_a2.outcome</code></li>
<li><code>mr_keep.outcome</code></li>
<li><code>data_source.outcome</code></li>
</ul>
</div>
<div class="section level2">
<h2 id="more-advanced-use-of-local-data">More advanced use of local data<a class="anchor" aria-label="anchor" href="#more-advanced-use-of-local-data"></a>
</h2>
<p>We have developed a summary data format called “GWAS VCF”, which is
designed to store GWAS results in a strict and performant way. It is
possible to use this format with the TwoSampleMR package. Going down
this avenue also allows you to use LD proxy functionality using your own
LD reference files (or ones that we provide). For more details, see this
package that explains the format and how to query it in R:</p>
<p><a href="https://github.com/mrcieu/gwasvcf" class="external-link uri">https://github.com/mrcieu/gwasvcf</a></p>
<p>and this package for how to connect the data to other packages
including TwoSampleMR</p>
<p><a href="https://github.com/MRCIEU/gwasglue" class="external-link uri">https://github.com/MRCIEU/gwasglue</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
